diff --git a/Src/AIWNIOS_CodeGen.HC b/Src/AIWNIOS_CodeGen.HC
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -2027,6 +2027,7 @@ break;case IC_RIP:
     new=__HC_ICAdd_LNot(cc2);
     break;case IC_UNARY_MINUS:
     new=__HC_ICAdd_Neg(cc2);
+#ifdef TARGET_X86
     break;case IC_BSF:
     new=__HC_ICAdd_Bsf(cc2);
     break;case IC_BSR:
@@ -2067,6 +2068,30 @@ break;case IC_RIP:
     new=__HC_ICAdd_InvlPg(cc2);
     break;case IC_STRLEN:
     new=__HC_ICAdd_StrLen(cc2);
+#else
+    break;case IC_BSF:
+    case IC_BSR:
+    case IC_RDTSC:
+    case IC_GET_RFLAGS:
+    case IC_SET_RFLAGS:
+    case IC_SET_RAX:
+    case IC_SET_RBP:
+    case IC_SET_RSP:
+    case IC_IN_U8:
+    case IC_IN_U16:
+    case IC_IN_U32:
+    case IC_OUT_U8:
+    case IC_OUT_U16:
+    case IC_OUT_U32:
+    case IC_QUE_INIT:
+    case IC_QUE_INS:
+    case IC_QUE_INS_REV:
+    case IC_QUE_REM:
+    case IC_INVLPG:
+    case IC_STRLEN:
+      PrintErr("Intrinsic is not implemented for this target\n");
+      throw('Compiler');
+#endif
     break;case IC_SQRT:
     new=__HC_ICAdd_Sqrt(cc2);
     break;case IC_DEREF:
