diff --git a/Src/AIWNIOS_CodeGen.HC b/Src/AIWNIOS_CodeGen.HC
index 3c4801e..ea5165b 100644
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -2170,6 +2170,8 @@ break;case IC_ASM:
 			   if(mlst->flags&MLF_STATIC) {
 				 if(mlst->static_data) {
 					MemCpy(tmpgh->data+mlst->offset,mlst->static_data,mlst->member_class->size*mlst->dim.total_cnt);
+					Free(mlst->static_data);
+					mlst->static_data=NULL;
 				 }
 			   }
 			   mlst=mlst->next;
diff --git a/Src/PrsVar.HC b/Src/PrsVar.HC
index 5d74704..af87671 100644
--- a/Src/PrsVar.HC
+++ b/Src/PrsVar.HC
@@ -604,8 +604,6 @@ pvl_restart3:
 	    tmpm->use_cnt=0;
 	    cc->flags|=CCF_DONT_MAKE_RES;
 	  }
-	  if (cc->flags&CCF_AOT_COMPILE)
-	    Free(tmpm->static_data);
 	  break;
 	case PRS1B_LOCAL_VAR:
 	  if (mode&PRSF_UNION) {
