--- a/c/assemble
+++ b/c/assemble
@@ -68,10 +68,8 @@ void AssembleFile(void) {
             }
 #endif
             else
             {
-                sprintf(buf, GCC_BIN_DIR "gcc -xassembler -mmodule %s %s -c %s -o %s",
-                             (CODE32) ? "-apcs32 -apcsfpv3" : "",
-                             throwback ? "-mthrowback" : "",
-                             opt.sfile, opt.ofile);
+                sprintf(buf, GCC_BIN_DIR "asasm -o %s %s",
+                             opt.ofile, opt.sfile);
             }
         }

--- a/c/writefile
+++ b/c/writefile
@@ -405,7 +405,7 @@ static void asm_header(void) {
   if (! CODE64)
   {
     fprintf(file, "\tAREA\t|!Header$$code|, %s, READONLY%s%s\n\n",
-                   (CODE26) ? "CODE" : "CODE, A32bit",
+                   (CODE26 || opt.toolchain == tc_gcc) ? "CODE" : "CODE, A32bit",
                    (opt.apcs & APCS_REENTRANT) ? ", PIC, REENTRANT" : "",
-                   (opt.apcs & APCS_FPE3) ? ", FP3" : "");
+                   (opt.toolchain != tc_gcc && (opt.apcs & APCS_FPE3)) ? ", FP3" : "");
   }
@@ -2188,6 +2188,6 @@ static void errors(void) {
     /* Still needs to be flagged as CODE because it isn't copyable */
     fprintf(file, "\tAREA\t|!Header$$code2|, %s, READONLY%s%s\n\n",
-                   (CODE26) ? "CODE" : "CODE, A32bit",
+                   (CODE26 || opt.toolchain == tc_gcc) ? "CODE" : "CODE, A32bit",
                    (opt.apcs & APCS_REENTRANT) ? ", PIC, REENTRANT" : "",
-                   (opt.apcs & APCS_FPE3) ? ", FP3" : "");
+                   (opt.toolchain != tc_gcc && (opt.apcs & APCS_FPE3)) ? ", FP3" : "");
   }
