diff --git a/Makefile b/Makefile
index 2cfe3cc..2f0ff47 100644
--- a/Makefile
+++ b/Makefile
@@ -21,7 +21,13 @@ OBJS = Top CLib CMOS Debug Interrupts SDIO Timers UART USB Video DMA Messaging G
 
 include HAL
 
+ifeq (GNU,${TOOLCHAIN})
+ASASM_TOOLOPTIONS :=
+ASFLAGS += -PD "GNUObjAsm SETL {TRUE}"
+else
 CFLAGS += -APCS 3/32bit/nofp/noswst
+ASFLAGS += -PD "GNUObjAsm SETL {FALSE}"
-ASFLAGS += -APCS 3/nofp/noswst
+endif
+ASFLAGS += -APCS 3/nofp/noswst -cpu 7-A
 
 # Dynamic dependencies:
diff --git a/s/GPIO b/s/GPIO
index 8e9fece..ff6c44c 100644
--- a/s/GPIO
+++ b/s/GPIO
@@ -124,7 +124,7 @@ PinDef_Alts   SETA 0
 var     SETS "PinDef_List" :CC: (:STR:PinDef_List:RIGHT:1)
 $var    SETA ($num) :OR: ($alt :SHL: 8) :OR: ($type :SHL: 16)
 PinDef_List SETA PinDef_List + 1
-PinDef_Alts SETA PinDef_Alts :OR (1 :SHL: $alt)
+PinDef_Alts SETA PinDef_Alts :OR: (1 :SHL: $alt)
         MEND
 
         MACRO
diff --git a/s/Messaging b/s/Messaging
index 8eb3af6..f0666ee 100644
--- a/s/Messaging
+++ b/s/Messaging
@@ -117,7 +117,7 @@ ReadSafetyCatch ROUT
         ANDS    a1, a2, #BoardRevisionExt_CM4_No_eMMC
         B       %FT20
 10
-        ADR     a2, tagbcm3
+        ADRL    a2, tagbcm3
         MOV     a3, #tagslencm3
         BL      memcpy                         ; copy to workspace buffer
 
@@ -209,8 +209,8 @@ HAL_QueryPlatform    ROUT
         MOVEQ   a2, #0
 
         LDMNEIA a3, {a1, a2}           ; MAC 11:22:33:44:55:66 => a2=xxxx6655 a1=44332211
-        REV     a3, a1
-        REV     a2, a2
+        DCI     &E6BF2F30               ; REV a3, a1
+        DCI     &E6BF1F31               ; REV a2, a2
         MOV     a2, a2, LSR #8
         ORR     a1, a2, a3, LSL #24
         BIC     a1, a1, #&FF
@@ -592,4 +592,3 @@ BCMMBox_Sleep
         MOV     pc, lr
 
                 END
-
diff --git a/s/Top b/s/Top
index df70c37..890cc4e 100644
--- a/s/Top
+++ b/s/Top
@@ -267,7 +267,7 @@ start
         MSR     SPSR_cxsf, #F32_bit+I32_bit+SVC32_mode
         MSR     SPSR_x, #A32_bit
         MSR     elr_hyp, lr
-        ERET
+        DCI     &E160006E               ; ERET
 02
         MSR     CPSR_c,#F32_bit+I32_bit+SVC32_mode
 03
diff --git a/s/memcpy b/s/memcpy
index 0909070..8b3cd4f 100644
--- a/s/memcpy
+++ b/s/memcpy
@@ -48,7 +48,7 @@ $offset Copy1Word$cond  $r0, $r1
       |
         LSR$cond $r0, $r1, #$offset*8
         LDR$cond $r1, [a2, #4]!
-        ORR$cond $r0, $r1, LSL #32-$offset*8
+        ORR$cond $r0, $r0, $r1, LSL #32-$offset*8
       ]
         STR$cond $r0, [a1], #4
         MEND
@@ -63,9 +63,9 @@ $offset Copy2Words$cond  $r0, $r1, $r2
         LDR$cond $r1, [a2, #4]!
         LSR$cond $r0, $r2, #$offset*8
         LDR$cond $r2, [a2, #4]!
-        ORR$cond $r0, $r1, LSL #32-$offset*8
-        LSR$cond $r1, #$offset*8
-        ORR$cond $r1, $r2, LSL #32-$offset*8
+        ORR$cond $r0, $r0, $r1, LSL #32-$offset*8
+        LSR$cond $r1, $r1, #$offset*8
+        ORR$cond $r1, $r1, $r2, LSL #32-$offset*8
       ]
         STM$cond a1!, {$r0,$r1}
         MEND
@@ -83,13 +83,13 @@ $offset Copy4Words  $r0, $r1, $r2, $r3, $r4
         LDMIB   a2!, {$r1,$r2}
         LSR     $r0, $r4, #$offset*8
         LDMIB   a2!, {$r3,$r4}
-        ORR     $r0, $r1, LSL #32-$offset*8
-        LSR     $r1, #$offset*8
-        ORR     $r1, $r2, LSL #32-$offset*8
-        LSR     $r2, #$offset*8
-        ORR     $r2, $r3, LSL #32-$offset*8
-        LSR     $r3, #$offset*8
-        ORR     $r3, $r4, LSL #32-$offset*8
+        ORR     $r0, $r0, $r1, LSL #32-$offset*8
+        LSR     $r1, $r1, #$offset*8
+        ORR     $r1, $r1, $r2, LSL #32-$offset*8
+        LSR     $r2, $r2, #$offset*8
+        ORR     $r2, $r2, $r3, LSL #32-$offset*8
+        LSR     $r3, $r3, #$offset*8
+        ORR     $r3, $r3, $r4, LSL #32-$offset*8
       ]
         STM     a1!, {$r0,$r1,$r2,$r3}
         MEND
@@ -117,21 +117,21 @@ $offset Copy8Words  $r0, $r1, $r2, $r3, $r4, $r5, $r6, $r7, $r8, $pld_dist
       [ "$pld_dist"<>""
         PLD     [a2, $pld_dist]
       ]
-        ORR     $r0, $r1, LSL #32-$offset*8
-        LSR     $r1, #$offset*8
-        ORR     $r1, $r2, LSL #32-$offset*8
-        LSR     $r2, #$offset*8
-        ORR     $r2, $r3, LSL #32-$offset*8
-        LSR     $r3, #$offset*8
-        ORR     $r3, $r4, LSL #32-$offset*8
-        LSR     $r4, #$offset*8
-        ORR     $r4, $r5, LSL #32-$offset*8
-        LSR     $r5, #$offset*8
-        ORR     $r5, $r6, LSL #32-$offset*8
-        LSR     $r6, #$offset*8
-        ORR     $r6, $r7, LSL #32-$offset*8
-        LSR     $r7, #$offset*8
-        ORR     $r7, $r8, LSL #32-$offset*8
+        ORR     $r0, $r0, $r1, LSL #32-$offset*8
+        LSR     $r1, $r1, #$offset*8
+        ORR     $r1, $r1, $r2, LSL #32-$offset*8
+        LSR     $r2, $r2, #$offset*8
+        ORR     $r2, $r2, $r3, LSL #32-$offset*8
+        LSR     $r3, $r3, #$offset*8
+        ORR     $r3, $r3, $r4, LSL #32-$offset*8
+        LSR     $r4, $r4, #$offset*8
+        ORR     $r4, $r4, $r5, LSL #32-$offset*8
+        LSR     $r5, $r5, #$offset*8
+        ORR     $r5, $r5, $r6, LSL #32-$offset*8
+        LSR     $r6, $r6, #$offset*8
+        ORR     $r6, $r6, $r7, LSL #32-$offset*8
+        LSR     $r7, $r7, #$offset*8
+        ORR     $r7, $r7, $r8, LSL #32-$offset*8
     ]
         STM     a1!, {$r0,$r1,$r2,$r3}
         STM     a1!, {$r4,$r5,$r6,$r7}
@@ -141,14 +141,14 @@ $offset Copy8Words  $r0, $r1, $r2, $r3, $r4, $r5, $r6, $r7, $r8, $pld_dist
         MACRO
 $offset Leading15Bytes
         LSLS    a4, ip, #31
-        SUB     a3, ip
+        SUB     a3, a3, ip
         LDRBMI  a4, [a2], #1
       [ $offset:AND:1 = 0
         LDRHCS  v1, [a2], #2
       |
         LDRBCS  v1, [a2], #1
         LDRBCS  v2, [a2], #1
-        ORRCS   v1, v2, LSL #8
+        ORRCS   v1, v1, v2, LSL #8
       ]
         STRBMI  a4, [a1], #1
         STRHCS  v1, [a1], #2
@@ -163,19 +163,19 @@ $offset Copy2WordsCS a4, v1, lr
 
         MACRO
 $offset Trailing15Bytes
-        LSLS    a3, #29
+        LSLS    a3, a3, #29
 $offset Copy2WordsCS a4, v1, lr
 $offset Copy1WordMI a4, lr
       [ $offset>0
-        ADD     a2, #$offset
+        ADD     a2, a2, #$offset
       ]
-        LSLS    a3, #2
+        LSLS    a3, a3, #2
       [ $offset:AND:1 = 0
         LDRHCS  a4, [a2], #2
       |
         LDRBCS  a4, [a2], #1
         LDRBCS  v1, [a2], #1
-        ORRCS   a4, v1, LSL #8
+        ORRCS   a4, a4, v1, LSL #8
       ]
         LDRBMI  v1, [a2]
         STRHCS  a4, [a1], #2
@@ -188,29 +188,29 @@ $offset Large
 $offset
 $offset Leading15Bytes
         AND     ip, a2, #31             ; distance back to start of first (or only) cacheline loaded for this 32-byte block
-        RSB     ip, #4*32               ; offset by which to PLD after we've LDMed each block
+        RSB     ip, ip, #4*32           ; offset by which to PLD after we've LDMed each block
 10
 $offset Copy8Words a4, v1, v2, v3, v4, v5, v6, sl, lr, ip
-        SUBS    a3, #32
+        SUBS    a3, a3, #32
         BHS     %BT10
         ; There are three possibilities for the remainder of the copy, depending
         ; on the sum of the present offset of a2 from its cacheline base and the
         ; number of trailing <= 31 bytes: either 0, 1 or 2 further preloads
         AND     a4, a2, #31
         AND     v1, a3, #31
-        ADD     ip, #32
-        ADDS    a4, v1
+        ADD     ip, ip, #32
+        ADDS    a4, a4, v1
         BEQ     %FT20
         PLD     [a2, ip]
-        ADD     ip, #32
+        ADD     ip, ip, #32
         CMP     a4, #32
         BLS     %FT20
         PLD     [a2, ip]
 20      ; Readjust loop termination condition to stop when less than 32 bytes left
-        ADD     a3, #4*32
+        ADD     a3, a3, #4*32
 30
 $offset Copy8Words a4, v1, v2, v3, v4, v5, v6, sl, lr
-        SUBS    a3, #32
+        SUBS    a3, a3, #32
         BHS     %BT30
         ; Deal with trailing <= 31 bytes
         LSLS    a4, a3, #31-4
@@ -229,7 +229,7 @@ $offset
 $offset Leading15Bytes
 10
 $offset Copy4Words a4, v1, v2, ip, lr
-        SUBS    a3, #16
+        SUBS    a3, a3, #16
         BHS     %BT10
         TST     a3, #15                 ; unaffected by initial adjustment by 16
         BEQ     done_smallframe
@@ -281,7 +281,7 @@ FastMemcpy ROUT
         ; Handle large copies
         PUSH    {a1,v1-v6,sl,lr}
         SUB     lr, a2, a1              ; find relative alignment
-        SUB     a3, #32+4*32            ; simplify first loop termination test
+        SUB     a3, a3, #32+4*32        ; simplify first loop termination test
                                         ; (we want it to stop when there are
                                         ; still 4 lots of 32 bytes to go)
         BIC     a4, a2, #31             ; -> base of first cacheline
@@ -295,9 +295,9 @@ o       SETA    o + 1
         ; If the leading up-to-15 bytes (to acquire destination alignment)
         ; will push the source pointer into a new cacheline, one more is needed
         ANDS    ip, a1, #15
-        RSBNE   ip, #16                 ; number of leading bytes
+        RSBNE   ip, ip, #16             ; number of leading bytes
         ADD     v1, a2, ip
-        AND     v1, #31                 ; index into cacheline when leading bytes done
+        AND     v1, v1, #31             ; index into cacheline when leading bytes done
         CMP     ip, v1                  ; preload needed if leading bytes exceed that
         BLS     %F10
         PLD     [a4, #4*32]
@@ -318,8 +318,8 @@ do_small ; Handle copies that write half-cachelines without preload
         SUB     lr, a2, a1              ; find relative alignment
         SUB     a3, a3, #16             ; simplify loop termination test
         ANDS    ip, a1, #15
-        RSBNE   ip, #16                 ; number of leading bytes
-        LSLS    lr, #31
+        RSBNE   ip, ip, #16             ; number of leading bytes
+        LSLS    lr, lr, #31
         BHI     %FA3
         BCS     %FA2
         BNE     %FA1
