diff --git a/hdr/CastleMacros b/hdr/CastleMacros
--- a/hdr/CastleMacros
+++ b/hdr/CastleMacros
@@ -16,7 +16,12 @@
         MACRO
         HALEntry $name
         ASSERT  (. - $table) / 4 = EntryNo_$name
+      [ GNUObjAsm
+        DCD     $name + (. - $table)
+        RELOC   R_ARM_REL32
+      |
         DCD     $name - $table
+      ]
         MEND
 
         MACRO
@@ -23,15 +28,35 @@
         VarEntry $name
         ASSERT  (. - $table) / 4 = EntryNo_$name
       [ $table_idx == 0
+      [ GNUObjAsm
+        DCD     ARM11_$name + (. - $table)
+        RELOC   R_ARM_REL32
+      |
         DCD     ARM11_$name - $table
+      ]
       ELIF $table_idx == 1
+      [ GNUObjAsm
+        DCD     QA7_$name + (. - $table)
+        RELOC   R_ARM_REL32
+      |
         DCD     QA7_$name - $table
+      ]
       |
+      [ GNUObjAsm
+        DCD     VC6_$name + (. - $table)
+        RELOC   R_ARM_REL32
+      |
         DCD     VC6_$name - $table
+      ]
       ]
         MEND
 
         MACRO
         NullEntry
+      [ GNUObjAsm
+        DCD     HAL_Null + (. - $table)
+        RELOC   R_ARM_REL32
+      |
         DCD     HAL_Null - $table
+      ]
         MEND
