diff --git a/Src/KernelA.HH b/Src/KernelA.HH
index 66b9dac..1f6752e 100644
--- a/Src/KernelA.HH
+++ b/Src/KernelA.HH
@@ -1530,9 +1530,9 @@ public class CAutoCompleteDictGlbls
 #define OPTf_KEEP_RELOCATIONS 0x26 //Added by nroot,will keep the CHashImport's
 #define OPTf_BOUNDS_CHECK 0x27 //Added by nroot,Bounds checker
 #define OPTf_BYTE_CODE 0x28 //Added by nroot
-#define OPTf_TEMPLEOS_BIN_SIGNATURE 0x2A
-#define OPTf_X86_SSE_RUNTIME 0x2B
 #define OPTf_RAW_AOT 0x29
+#define OPTf_TEMPLEOS_BIN_SIGNATURE 0x2A
+#define OPTf_X86_SSE_RUNTIME 0x2B
 
 #define OPTF_ECHO		(1<<OPTf_ECHO)
 
@@ -2173,8 +2173,8 @@ class CLexHashTableContext
 #define CCF_PAREN		0x2000000000
 #define CCf_PAREN		37
 #define CCF_CLASS_DOL_OFFSET	0x4000000000
-#define CCF_RAW_AOT_ASM		0x10000000000
 #define CCF_DONT_MAKE_RES	0x8000000000
+#define CCF_RAW_AOT_ASM		0x10000000000
 
 public class CCmpCtrl
 {
diff --git a/Src/UAsmX64.HC b/Src/UAsmX64.HC
index 40896d2..f7d264a 100644
--- a/Src/UAsmX64.HC
+++ b/Src/UAsmX64.HC
@@ -1322,6 +1322,7 @@ U0 UAsmInit() {
   AO("FSTP",-1,-1,0xdd,-1,X86_OPCF_MODRM,X86_OPERAND_M64,X86_OPERAND_EXT,3);
   AO("FSTP",-1,-1,0xdd,-1,X86_OPCF_MODRM,X86_OPERAND_STi,X86_OPERAND_EXT,3);
   AO("FSTSW",0x9b,-1,0xdd,-1,X86_OPCF_MODRM,X86_OPERAND_M16,X86_OPERAND_EXT,7);
+  //TempleOS uses bare FSTSW as the implicit AX form.
   AO("FSTSW",0x9b,-1,0xdf,0xe0,0);
   AO("FSUB",-1,-1,0xd8,-1,X86_OPCF_MODRM,X86_OPERAND_ST0,X86_OPERAND_M32,X86_OPERAND_EXT,4);
   AO("FSUB",-1,-1,0xd8,-1,X86_OPCF_MODRM,X86_OPERAND_ST0,X86_OPERAND_STi,X86_OPERAND_EXT,4);
@@ -1688,7 +1689,8 @@ U0 UAsmInit() {
 //POP FS
   AO("POPCNT",0xf3,0xf,0xb8,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_RMx);
   AO("POPF",-1,-1,0x9d,-1,0);
-  AO("POPFW,POPFD",-1,-1,0x9d,-1,0);
+  AO("POPFW",-1,-1,0x9d,-1,X86_OPCF_OPSIZE16);
+  AO("POPFD",-1,-1,0x9d,-1,0);
   AO("POPAW,POPAD",-1,-1,0x61,-1,0);
   AO("POR",-1,0xf,0xeb,-1,X86_OPCF_MODRM,X86_OPERAND_RMM,X86_OPERAND_RMMM64);
   AO("POR",0x66,0xf,0xeb,-1,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM);
@@ -1780,7 +1782,8 @@ U0 UAsmInit() {
 //PUSH FS
 //PUSH Gs
   AO("PUSHFQ,PUSHF",-1,-1,0x9c,-1,0);
-  AO("PUSHFW,PUSHFD",-1,-1,0x9c,-1,0);
+  AO("PUSHFW",-1,-1,0x9c,-1,X86_OPCF_OPSIZE16);
+  AO("PUSHFD",-1,-1,0x9c,-1,0);
   AO("PUSHAW,PUSHAD",-1,-1,0x60,-1,0);
 
   AO("PXOR",-1,0xf,0xef,-1,X86_OPCF_MODRM,X86_OPERAND_RMM,X86_OPERAND_RMMM64);
