diff --git a/Src/CMain.HC b/Src/CMain.HC
index 2ae90a8..d158dc6 100644
--- a/Src/CMain.HC
+++ b/Src/CMain.HC
@@ -413,6 +413,8 @@ I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
 #ifdef TARGET_RISCV
     bfh->bin_signature='RV64';
 #endif
+    if (GetOption(OPTf_TEMPLEOS_BIN_SIGNATURE))
+      bfh->bin_signature=BIN_SIGNATURE_VAL;
     MemCpy(bfh(U8 *)+sizeof(CBinFile),tmpaot->buf,aot_U8s);
     MemCpy(bfh(U8 *)+sizeof(CBinFile)+aot_U8s,patch_table,
 	  size-aot_U8s-sizeof(CBinFile));
diff --git a/Src/KernelA.HH b/Src/KernelA.HH
index e9624fc..66b9dac 100644
--- a/Src/KernelA.HH
+++ b/Src/KernelA.HH
@@ -1530,6 +1530,7 @@ public class CAutoCompleteDictGlbls
 #define OPTf_KEEP_RELOCATIONS 0x26 //Added by nroot,will keep the CHashImport's
 #define OPTf_BOUNDS_CHECK 0x27 //Added by nroot,Bounds checker
 #define OPTf_BYTE_CODE 0x28 //Added by nroot
+#define OPTf_TEMPLEOS_BIN_SIGNATURE 0x2A
 #define OPTf_RAW_AOT 0x29
 
 #define OPTF_ECHO		(1<<OPTf_ECHO)
