diff --git a/c/arm_backend.c b/c/arm_backend.c
--- a/c/arm_backend.c
+++ b/c/arm_backend.c
@@ -3695,5 +3695,11 @@ static int64_t __OptPassFinal(CCmpCtrl *cctrl, CRPN *rpn, char *bin,
       code_off = ICMov(cctrl, &rpn->res, &next->res, bin, code_off);
       break;
+    case IC_IREG:
+      code_off = __OptPassFinal(cctrl, next, bin, code_off);
+      tmp = next->res;
+      tmp.raw_type = RT_PTR;
+      code_off = ICMov(cctrl, &rpn->res, &tmp, bin, code_off);
+      break;
     case IC_BASE_PTR:
       if (rpn->res.mode == MD_REG)
         into_reg = rpn->res.reg;
diff --git a/c/riscv_backend.c b/c/riscv_backend.c
--- a/c/riscv_backend.c
+++ b/c/riscv_backend.c
@@ -3748,5 +3748,11 @@ static int64_t __OptPassFinal(CCmpCtrl *cctrl, CRPN *rpn, char *bin,
       code_off = ICMov(cctrl, &rpn->res, &next->res, bin, code_off);
       break;
+    case IC_IREG:
+      code_off = __OptPassFinal(cctrl, next, bin, code_off);
+      tmp = next->res;
+      tmp.raw_type = RT_PTR;
+      code_off = ICMov(cctrl, &rpn->res, &tmp, bin, code_off);
+      break;
     case IC_BASE_PTR:
       if (rpn->res.mode == MD_REG)
         into_reg = rpn->res.reg;
diff --git a/c/x86_64_backend.c b/c/x86_64_backend.c
--- a/c/x86_64_backend.c
+++ b/c/x86_64_backend.c
@@ -7466,5 +7466,11 @@ static int64_t __OptPassFinal(CCmpCtrl *cctrl, CRPN *rpn, char *bin,
       code_off = ICMov(cctrl, &rpn->res, &next->res, bin, code_off);
       break;
+    case IC_IREG:
+      code_off = __OptPassFinal(cctrl, next, bin, code_off);
+      tmp = next->res;
+      tmp.raw_type = RT_PTR;
+      code_off = ICMov(cctrl, &rpn->res, &tmp, bin, code_off);
+      break;
     case IC_BASE_PTR:
       if (rpn->res.mode == MD_REG)
         into_reg = rpn->res.reg;
