diff --git a/Src/AIWNIOS_CodeGen.HC b/Src/AIWNIOS_CodeGen.HC
index 81f3104..1a8aece 100644
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -1911,7 +1911,7 @@ break;case IC_RIP:
           h(CHashFun*)->exe_addr
         );
       }
-    } else if (!(h->type&HTF_RESOLVE) ||
+    } else if (cc->flags&CCF_AOT_COMPILE || !(h->type&HTF_RESOLVE) ||
           h->type&HTF_UNRESOLVED) {
         new=__HC_ICAdd_Reloc(
             acc,
diff --git a/Src/AIWNIOS_PrsExp.HC b/Src/AIWNIOS_PrsExp.HC
index fe0bee9..d06cf9d 100644
--- a/Src/AIWNIOS_PrsExp.HC
+++ b/Src/AIWNIOS_PrsExp.HC
@@ -1745,12 +1745,21 @@ ins_label:
 	      rpn->ic_dim=cc->hash_entry(CHashGlblVar*)->dim.next;
 	      Lex(cc);
 	    } else if(cc->hash_entry->type&HTT_EXPORT_SYS_SYM) {
-	        rpn=RPNAddNew(
-		      cc,
-		      IC_IMM_I64,
-		      cc->hash_entry(CHashExport*)->val,
-		      cmp.internal_types[RT_I64],
-		      );
+              if (cc->flags&CCF_AOT_COMPILE &&
+                    !(cc->hash_entry->type&HTF_IMM))
+                rpn=RPNAddNew(
+                      cc,
+                      IC_ADDR_IMPORT,
+                      CodeMiscHashNew(cc,cc->hash_entry),
+                      cmp.internal_types[RT_I64],
+                      );
+              else
+                rpn=RPNAddNew(
+                      cc,
+                      IC_IMM_I64,
+                      cc->hash_entry(CHashExport*)->val,
+                      cmp.internal_types[RT_I64],
+                      );
 	      Lex(cc);
 	    } else if(cc->hash_entry->type&HTT_KEYWORD) {
 	      switch(cc->hash_entry(CHashGeneric *)->user_data0) {
