diff --git a/Src/CMain.HC b/Src/CMain.HC
index 6eb8c9d..de434d1 100644
--- a/Src/CMain.HC
+++ b/Src/CMain.HC
@@ -561,10 +561,36 @@ I64 StreamExePrint(U8 *fmt,...)
   return res;
 }
 
+CHash *AOTResolvedHashFind(CCmpCtrl *cc,U8 *name)
+{
+  CHash *h;
+  CHashTable *table=cc->htc.glbl_hash_table;
+  I64 i;
+  for (i=0;i<=table->mask;i++)
+    for (h=table->body[i];h;h=h->next)
+      if (h->type&HTT_FUN && !StrCmp(h->str,name) &&
+            h->type&HTF_RESOLVE &&
+            0<=h(CHashFun*)->exe_addr<cc->aotc->rip)
+        return h;
+  for (i=0;i<=table->mask;i++)
+    for (h=table->body[i];h;h=h->next)
+      if (h->type&HTT_GLBL_VAR && !StrCmp(h->str,name) &&
+            !(h(CHashGlblVar*)->flags&GVF_EXTERN))
+        return h;
+  for (i=0;i<=table->mask;i++)
+    for (h=table->body[i];h;h=h->next)
+      if (h->type&HTT_EXPORT_SYS_SYM && !StrCmp(h->str,name) &&
+            h->type&HTF_RESOLVE)
+        return h;
+  return NULL;
+}
+
 //Added by nroot
 U0 FillInShortCalls(CCmpCtrl *cc) {
   CAOTImportExport *tmpaotie,*tmpaotie1;
+  CAOTAbsAddr *tmpabs;
   CHashFun *fun;
+  Bool value_resolved;
   I64 i,ptr,low12;
   for(tmpaotie=cc->aot->next_ie;tmpaotie!=&cc->aot->next_ie;tmpaotie=tmpaotie1) {
     tmpaotie1=tmpaotie->next;
@@ -636,6 +662,36 @@ skip:;
         Free(tmpaotie);
 #endif
       }
+    } else if (tmpaotie->type==IET_IMM_I64 && tmpaotie->str) {
+      value_resolved=TRUE;
+      fun=AOTResolvedHashFind(cc,tmpaotie->str);
+      if (fun && fun->type&HTT_FUN && fun->type&HTF_RESOLVE)
+        i=fun->exe_addr+tmpaotie->offset;
+      else if (fun && fun->type&HTT_GLBL_VAR && fun->type&HTF_EXPORT &&
+            !(fun(CHashGlblVar*)->flags&(GVF_EXTERN|GVF_DATA_HEAP)))
+        i=fun(CHashGlblVar*)->data_addr_rip+tmpaotie->offset;
+      else if (fun && fun->type&HTT_EXPORT_SYS_SYM &&
+            fun->type&HTF_RESOLVE)
+        i=fun(CHashExport*)->val+tmpaotie->offset;
+      else
+        value_resolved=FALSE;
+      if (value_resolved) {
+        ptr=tmpaotie->rip;
+        if (tmpaotie->aot)
+          ptr+=tmpaotie->aot->rip2;
+        AOTStoreCodeU8At(cc,ptr,
+              i,i>>8,i>>16,i>>24);
+        AOTStoreCodeU8At(cc,ptr+4,
+              i>>32,i>>40,i>>48,i>>56);
+        tmpabs=CAlloc(sizeof(CAOTAbsAddr));
+        tmpabs->next=cc->aotc->abss;
+        tmpabs->type=AAT_ADD_U64;
+        tmpabs->rip=ptr;
+        cc->aotc->abss=tmpabs;
+        QueRem(tmpaotie);
+        Free(tmpaotie->str);
+        Free(tmpaotie);
+      }
     }
   }
 }
diff --git a/Src/CExts.HC b/Src/CExts.HC
index 296d3ae..89d87aa 100644
--- a/Src/CExts.HC
+++ b/Src/CExts.HC
@@ -42,2 +42,3 @@ extern U0 PrsVarInit2(CCmpCtrl *cc,U8 **_dst,CHashClass *tmpc,
 extern U0 FillInShortCalls(CCmpCtrl *cc);
+extern CHash *AOTResolvedHashFind(CCmpCtrl *cc,U8 *name);
 extern CHashClass *AssignRawTypeToNode(CRPN *r);
