diff --git a/c/x86_64_backend.c b/c/x86_64_backend.c
index 1c0b348..6edd785 100644
--- a/c/x86_64_backend.c
+++ b/c/x86_64_backend.c
@@ -9099,7 +9099,10 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
           code_off += 8 - code_off % 8;
         misc->addr = bin + code_off;
         if (bin)
-          *(void **)((char *)MemGetWritePtr(bin) + code_off) = &DoNothing;
+          *(void **)((char *)MemGetWritePtr(bin) + code_off) =
+              cctrl->flags & CCF_AOT_COMPILE
+                  ? NULL
+                  : &DoNothing;
         code_off += 8;
         if (run == 2) {
           *(import = A_CALLOC(sizeof(CHashImport), NULL)) = (CHashImport){
