diff --git a/s/KbdScan b/s/KbdScan
index e17a128..f42193e 100644
--- a/s/KbdScan
+++ b/s/KbdScan
@@ -36,19 +36,29 @@
 
         EXPORT  HAL_KbdScanDependencies
 
+      [ :DEF: Headless
+HAL_KbdScanDependencies ROUT
+        ADR     a1, %FT10
+        MOV     pc, lr
+10
+        DCB     "SharedCLibrary,BufferManager,DeviceFS,SerialDeviceSupport,"
+        DCB     "FileCore,SystemDevices,SDIODriver,SDFS", 0
+        ALIGN
+      |
 HAL_KbdScanDependencies ROUT
         CPUDetect a1
         ADRLS   a1, %FT10
         ADRHI   a1, %FT20
         MOV     pc, lr
 10
         ; Modules needed for keyboard scanning, no need to list those
         ; before 'FirstUnpluggableModule' since they can't be unplugged
         DCB     "SharedCLibrary,BufferManager,DeviceFS,RTSupport,USBDriver,"
         DCB     "DWCDriver,InternationalKeyboard", 0
 20
         ; Two USB controllers on Pi 4
         DCB     "SharedCLibrary,BufferManager,DeviceFS,RTSupport,USBDriver,"
         DCB     "DWCDriver,XHCIDriver,InternationalKeyboard", 0
+      ]
 
         END
diff --git a/s/UART b/s/UART
index 512c598..c152dcb 100644
--- a/s/UART
+++ b/s/UART
@@ -829,8 +829,12 @@ HAL_UARTModemStatus
         ; Ensure DSR
         ORR     a1, a1, #1:SHL:5
  |
+      [ :DEF: Headless
+        MOV     a1, #2_10110000
+      |
         ; Fake control lines as though a modem is plugged in but offline
         MOV     a1, #2_00110000
+      ]
  ]
         MOV     pc, lr
 
