diff --git a/s/NewReset b/s/NewReset
--- a/s/NewReset
+++ b/s/NewReset
@@ -424,8 +424,10 @@
         DebugTX "OscliInit"
         BL      OscliInit
 
+      [ :LNOT: :DEF: Headless
         BL      InitialiseMode          ; select correct screen mode, in case any
                                         ; module prints anything in initialisation
+      ]
 
         DebugTX "HAL_InitDevices"
         AddressHAL
@@ -468,2 +468,5 @@
         LDR     r3, [r3, #MetroGnome]
         ADD     r3, r3, #300            ; timeout (cs)
+      [ :DEF: Headless
+        B       %FT04
+      ]
@@ -513,6 +518,9 @@
         MOV     r6, r0
         CMP     r6, #-1
         BNE     %BT02                   ; the show's not over until the RAM clear is done
+      [ :DEF: Headless
+        B       %FT05
+      ]
 
         CMP     r5, #-1
         BEQ     %FT05                   ; HAL gave no KbdScan details
diff --git a/s/PMF/osinit b/s/PMF/osinit
--- a/s/PMF/osinit
+++ b/s/PMF/osinit
@@ -96,7 +96,9 @@ BuffPtrInitLoop
         MOV     R0, #4_3330             ; Assume VGA during osinit
         STRB    R0, [R1, #MonitorLeadType]
 
+      [ :LNOT: :DEF: Headless
         BL      ReadUniqueID
+      ]
 
         Push    "r9,r12"
         AddressHAL
@@ -116,8 +118,10 @@ secs0070 *      (86400*(365*70+18))
         BL      Store5ByteInRealTime
 
 ; insert soft key 10
+      [ :LNOT: :DEF: Headless
         MOV     R2, #&CA
         BL      RDCHS
+      ]
         Pull    PC
 
         LTORG
diff --git a/s/ModHand b/s/ModHand
--- a/s/ModHand
+++ b/s/ModHand
@@ -154,5 +154,8 @@ ModuleInitForKbdScan Entry "r0-r12"
         BNE     %FT40                                   ; if your name's not on the list you can't come in
 30
+      [ :DEF: Headless
+        MOV     r11, #0
+      |
         ADR     r11, UnplugCMOSTable
         SUBS    r14, r10, #FirstUnpluggableModule       ; subtract number of first module that has an unplug bit
         MOVCS   r1, r14, LSR #3                         ; get byte number
@@ -164,6 +167,7 @@ ModuleInitForKbdScan Entry "r0-r12"
         MOVCS   r12, #1
         ORRCS   r11, r11, r12, LSL r14                  ; merge with bit mask
         MOVCC   r11, #0                                 ; otherwise zero
+      ]
 
         BL      AddROMModuleNode
         BVS     %FT50                                   ; if failed then can't add any more ROMs!

@@ -240,6 +244,9 @@ ModuleInit   Entry "r0-r12"
         ADDEQ   r5, r6, #Module_HelpStr                 ; then use help offset as string (null string)
         ADDNE   r5, r6, r5                              ; otherwise point to help string
 
+      [ :DEF: Headless
+        MOV     r11, #0
+      |
         ADR     r11, UnplugCMOSTable
         SUBS    r14, r10, #FirstUnpluggableModule       ; subtract number of first module that has an unplug bit
         MOVCS   r1, r14, LSR #3                         ; get byte number
@@ -249,6 +256,7 @@ ModuleInit   Entry "r0-r12"
         MOVCS   r12, #1
         ORRCS   r11, r11, r12, LSL r14                  ; merge with bit mask
         MOVCC   r11, #0                                 ; otherwise zero
+      ]
 
         BL      AddROMModuleNode
         BVS     %FT50                                   ; if failed then can't add any more ROMs!
@@ -279,6 +287,9 @@ ModuleInit   Entry "r0-r12"

 ; now do podule ROMs

+      [ :DEF: Headless
+        B       %FT50
+      ]
 20
         MOV     r3, #0                                  ; start at podule 0
 21

diff --git a/s/NewReset b/s/NewReset
--- a/s/NewReset
+++ b/s/NewReset
@@ -658,6 +658,7 @@ init_other_modules
         MOV     R0, #FSControl_SelectFS ; set configured filing system
         SWI     XOS_FSControl
 
+      [ :LNOT: :DEF: Headless
         ; Update territory in case it changed after ModuleInitForKbdScan
         MOV     R0, #ReadCMOS
         MOV     R1, #TerritoryCMOS
@@ -667,6 +668,7 @@ init_other_modules
 
         ; Update RTC now all the modules are running
         SWI     XOS_ResyncTime
+      ]
 
   [ DebugROMInit
         SWI     XOS_WriteS
@@ -682,6 +684,9 @@ init_other_modules
   ]
         SWI     XOS_LeaveOS
         SWI     XOS_EnterOS             ; switch back to SVC mode (IRQs, FIQs enabled)
+      [ :DEF: Headless
+        B       DoStartSuper
+      ]
 
       [ International                   ; Open the kernel messages file.
         LDR     r0, =ZeroPage+KernelMessagesBlock+4
@@ -718,7 +718,9 @@ init_other_modules
         SWI     XOS_Memory
       ]
 
+      [ :LNOT: :DEF: Headless
         BL      InitialiseMode
+      ]
         LDR     R0, =ZeroPage
         LDRB    R14, [R0, #MentionCMOSReset]
         TEQ     R14, #0
@@ -981,13 +981,21 @@ DoStartSuper
         CMP     R2, #0
         ADDNE   R1, R3, R2
 DoStartSuper
+      [ :DEF: Headless
+        ADRL    R1, MountinLauncherTitle
+      |
         ADREQL  R1, UtilModTitle        ; ALWAYS enter via SWI: sets CAO etc.
+      ]
         MOV     R0, #ModHandReason_Enter
         ADRL    R2, crstring            ; no environment
         SWI     XOS_Module
         CMP     r0, r0                  ; set EQ if failed to enter config.lang
         B       DoStartSuper            ; -> force Super entry
 
+MountinLauncherTitle
+        =       "MountinLauncher", 0
+        ALIGN
+
 
 str_booterror   DCB     "Boot$$Error",0
                 ALIGN
