--- a/Src/MakeISOFile.HC
+++ b/Src/MakeISOFile.HC
@@ -34,6 +34,8 @@
   U64 BDVD_BLK_CNT=0x8f46;
   U64 BDVD_SHIFT_BLKS=0x8f48;
 
+  U32 stage2_blk;
+  U16 stage2_cnt,stage2_shift;
   CDirEntry de;
   CFile *out_file=NULL;
   U8	*stage1_buf=CAlloc(DVD_BOOT_LOADER_SIZE);
@@ -41,10 +43,12 @@
   if (FileFind(stage2_filename,&de) && (out_file=FOpen(iso_filename,"wc+"))) {
     MemCpy(stage1_buf,stage1_src,BDVD_END-BDVD_START);
-    *(BDVD_BLK_LO	-BDVD_START+stage1_buf)(U32 *)=de.clus>>2;
-    *(BDVD_BLK_CNT	-BDVD_START+stage1_buf)(U16 *)=
-	  (de.size+DVD_BLK_SIZE-1)>>(BLK_SIZE_BITS+2);
-    *(BDVD_SHIFT_BLKS	-BDVD_START+stage1_buf)(U16 *)=de.clus&3;
-    if (de.clus&3)
-      *(BDVD_BLK_CNT	-BDVD_START+stage1_buf)(U16 *)+=1;
+    stage2_blk=de.clus>>2;
+    stage2_cnt=(de.size+DVD_BLK_SIZE-1)>>(BLK_SIZE_BITS+2);
+    stage2_shift=de.clus&3;
+    if (stage2_shift)
+      stage2_cnt++;
+    MemCpy(stage1_buf+BDVD_BLK_LO-BDVD_START,&stage2_blk,sizeof(U32));
+    MemCpy(stage1_buf+BDVD_BLK_CNT-BDVD_START,&stage2_cnt,sizeof(U16));
+    MemCpy(stage1_buf+BDVD_SHIFT_BLKS-BDVD_START,&stage2_shift,sizeof(U16));
     FBlkWrite(out_file,stage1_buf,
 	  20<<2+1<<2,DVD_BOOT_LOADER_SIZE/BLK_SIZE);
@@ -68,7 +72,7 @@
   CFile *out_file=NULL;
 
   if (out_file=iso_filename) {
-    iso_size=FSize(out_file)/DVD_BLK_SIZE;
+    iso_size=(dv->bd->max_blk+1)/(DVD_BLK_SIZE/BLK_SIZE);
     for (i=0;i<dv->bd->drv_offset;i+=4)
       FBlkWrite(out_file,zero_buf,i,4);
