--- a/Src/AsmResolve.HC
+++ b/Src/AsmResolve.HC
@@ -32,8 +32,7 @@
 {
   CAOTImportExport *tmpie;
   CAsmUndefHash *tmpauh;
-  I64 (*machine_code)()=tmpu->machine_code;
-  I64 res=(*machine_code)();
+  I64 res=Call(tmpu->machine_code);
   if (undefined) {
     tmpauh=tmpu->asm_undef_hash;
     while (tmpauh) {
--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -1243,7 +1243,6 @@
   CHashExport *exp;
   CCodeMisc *cm;
   Bool local;
-  I64 (*machine_code)();
   patch=patches->next;
   while(patch!=patches) {
     if(!patch->imm) goto nxt;
@@ -1306,9 +1305,8 @@
 resolved:
       cur=cur->next;
     }
-    machine_code=patch->imm->machine_code;
-    patch->imm->off=(*machine_code)();
-    Free(patch->imm->machine_code);
+    patch->imm->off=Call(patch->imm->machine_code);
+    Free(patch->imm->machine_code);
     patch->imm->machine_code=NULL;
     LinkedLstDel(tab);
 nxt:
