diff --git a/Src/AsmX64.HC b/Src/AsmX64.HC
index b48d0cd..de2da3a 100644
--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -1497,7 +1497,25 @@ enter:
       seg_size=64;
       Lex(cc);
     } else if(!StrCmp(cc->hash_entry->str,"IMPORT")) {
-//TODO
+      Lex(cc);
+      while(cc->token&&cc->token!=';') {
+        if(cc->token!=TK_IDENT)
+          LexExcept(cc,"Expected identifier at");
+        exp=CAlloc(sizeof CHashExport);
+        exp->str=cc->cur_str;
+        cc->cur_str=NULL;
+        exp->type=HTT_EXPORT_SYS_SYM|HTF_UNRESOLVED;
+        HashSrcFileSet(cc,exp);
+        HashAdd(exp,cc->htc.glbl_hash_table);
+        exp->type|=HTF_IMPORT;
+        if(Lex(cc)==',')
+          Lex(cc);
+        else if(cc->token!=';')
+          LexExcept(cc,"Expected ',' or ';' at");
+      }
+      if(cc->token!=';')
+        LexExcept(cc,"Missing ';' at");
+      Lex(cc);
     } else if(!StrCmp(cc->hash_entry->str,"BINFILE")) {
       if(Lex(cc)!=TK_STR)
         LexExcept(cc,"Expected a filename.");
