diff --git a/Src/AIWNIOS_CodeGen.HC b/Src/AIWNIOS_CodeGen.HC
index 5bc7be5..560b93a 100644
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -1406,7 +1406,8 @@ skip_convert:;
   if(cc->flags&CCF_AOT_COMPILE)
     AiwniosMakeShortJmps(cc,acc);
   #endif
-  if(cc->flags&CCF_AOT_COMPILE)
+  if (cc->flags&CCF_AOT_COMPILE &&
+        !(cc->flags&CCF_ASM_EXPRESSIONS))
     __HC_CmpCtrl_SetAOT(acc);
   if(cc->flags&CCF_RAW_AOT_ASM)
     __HC_CmpCtrl_SetRawAOTAsm(acc);
@@ -2210,7 +2211,8 @@ break;case IC_RIP:
       // So if we have relocations that aren't used (have an ->addr of INVALID_PTR)
       // dont add a "fill-in-later" for it
       //
-      if(cc->flags&CCF_AOT_COMPILE) {
+      if (cc->flags&CCF_AOT_COMPILE &&
+            !(cc->flags&CCF_ASM_EXPRESSIONS)) {
 		if(misc->addr!=INVALID_PTR) {
           //All non-imported functions are short calls
           if(misc->flags&CMF_SHORT_ADDR) {
