diff --git a/Src/AIWNIOS_PrsExp.HC b/Src/AIWNIOS_PrsExp.HC
index 5adeb1a..d806d18 100644
--- a/Src/AIWNIOS_PrsExp.HC
+++ b/Src/AIWNIOS_PrsExp.HC
@@ -833,6 +833,19 @@ CCodeMisc *CodeMiscHashNew(CCmpCtrl *cc,CHash* h) {
   return cur;
 }
 
+U0 AsmExpressionRelocAdd(CCmpCtrl *cc,CHash *h)
+{
+  CAsmUndefHash *tmpauh;
+
+  if (cc->flags&CCF_ASM_EXPRESSIONS) {
+    tmpauh=CAlloc(sizeof CAsmUndefHash);
+    tmpauh->hash=h;
+    tmpauh->next=cc->asm_undef_hash;
+    cc->asm_undef_hash=tmpauh;
+    cc->flags|=CCF_UNRESOLVED;
+  }
+}
+
 U0 DbgPrint(U8 *fmt,...) {
   U8 *s=StrPrintJoin(NULL,fmt,argc,argv);
   DbgPutS(s);
@@ -996,6 +1009,8 @@ Bool PrsExpression(CCmpCtrl *cc,U8 **ul=NULL,I64 flags=0,CPrsStk *_ps=NULL) { //
   CHashClass *tmpc;
   CMemberLst *mlst;
   CAsmUndefHash *tmpauh;
+  U8 *lookahead;
+  Bool aot_fun_import;
 //u32[0] is type
   //u32[1] is prec
   CRPN *rpn,*rpn2;
@@ -1459,8 +1474,14 @@ found_end_square:
 	  last_consumes_right=0;
 	  Lex(cc);
 	  break;case TK_IDENT:
+	  if(cc->flags&CCF_ASM_EXPRESSIONS&&
+        HashFind(cc->cur_str,cmp.asm_hash,HTT_OPCODE|HTT_ASM_KEYWORD)) {
+	    finish=TRUE;
+	    break;
+	  }
 	  if(finish) break;
 	  last_consumes_right=0;
+          aot_fun_import=FALSE;
 	  if(cc->flags&CCF_ASM_EXPRESSIONS) {
 	    if(cc->local_var_entry&&cc->htc.fun) {
 	      if(flags&PEF_ADDR_OF) {
@@ -1473,12 +1494,29 @@ found_end_square:
 	      tmpex->str=MStrPrint("%s.%s",cc->htc.fun->str,cc->cur_str);
 	      tmpex->type=HTT_EXPORT_SYS_SYM | HTF_UNRESOLVED;
 	      goto ins_label;
-	    } else if(cc->hash_entry&&cc->hash_entry->type&HTT_CLASS) {
+	    } else if(cc->aot&&cc->hash_entry&&
+                  cc->hash_entry->type&HTT_FUN&&
+                  !Bt(&cc->hash_entry(CHashFun*)->flags,Ff_INTERNAL)) {
+              aot_fun_import=TRUE;
+              cc->hash_entry=NULL;
+              goto new;
+            } else if(cc->hash_entry&&cc->hash_entry->type&HTT_CLASS) {
 //MOV RAX, CTask.addr[RBX]
-	      PrsOffsetOf(cc);
-	      break;
+              lookahead=cc->cur_buf_ptr;
+              while(*lookahead&&*lookahead<=' ') lookahead++;
+              if(cc->flags&CCF_USE_LAST_U16&&cc->last_U16=='.'||
+                    *lookahead=='.') {
+	        PrsOffsetOf(cc);
+	        break;
+              }
+              cc->hash_entry=NULL;
+              goto new;
 	    } else if(!cc->hash_entry) {
 new:
+	      if(flags&PEF_ADDR_OF) {
+	        PrsPop(cc->ps);
+	        RPNDel(PrsPop2(cc->ps));
+	      }
 	      tmpex=CAlloc(sizeof(CHashExport));
 	      tmpex->str=StrNew(cc->cur_str);
 	      if (*tmpex->str=='@' && tmpex->str[1]=='@') {
@@ -1494,11 +1532,13 @@ new:
 	      } else {
 	        tmpex->type=HTT_EXPORT_SYS_SYM | HTF_UNRESOLVED;
 ins_label:
-	        tmpc=HashFind(tmpex->str,cc->htc.glbl_hash_table,HTT_EXPORT_SYS_SYM);
-	        if (tmpc)
-		  HashAddAfter(tmpex,tmpc,cc->htc.glbl_hash_table);
-	        else
-		  HashAdd(tmpex,cc->htc.glbl_hash_table);
+                if (!aot_fun_import) {
+	          tmpc=HashFind(tmpex->str,cc->htc.glbl_hash_table,HTT_EXPORT_SYS_SYM);
+	          if (tmpc)
+		    HashAddAfter(tmpex,tmpc,cc->htc.glbl_hash_table);
+	          else
+		    HashAdd(tmpex,cc->htc.glbl_hash_table);
+                }
 	      }
 	      cc->flags|=CCF_UNRESOLVED;
 	      rpn=RPNAddNew(
@@ -1520,6 +1560,7 @@ ins_label:
 		    CodeMiscHashNew(cc,cc->hash_entry),
 		    cmp.internal_types[RT_I64],
 		    );
+	      AsmExpressionRelocAdd(cc,cc->hash_entry);
 	      Lex(cc);
 	      goto fin_ident;
 	    } else if(cc->hash_entry->type&HTT_EXPORT_SYS_SYM) {
@@ -1570,6 +1611,7 @@ ins_label:
 		      CodeMiscHashNew(cc,cc->hash_entry),
 		      cc->hash_entry,
 		      );
+	          AsmExpressionRelocAdd(cc,cc->hash_entry);
 		      #ifdef TARGET_RISCV
 		      rpn->ic_data(CCodeMisc*)->flags|=CMF_SHORT_ADDR;
 		      #endif
@@ -1652,6 +1694,7 @@ ins_label:
 		      CodeMiscHashNew(cc,cc->hash_entry),
 		      cc->hash_entry(CHashGlblVar*)->var_class,
 		      );
+	        AsmExpressionRelocAdd(cc,cc->hash_entry);
 	      } else {
 		  rpn=RPNAddNew(
 		      cc,
