--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -2251,2 +2251,12 @@
           } else if(misc->h->type&HTT_GLBL_VAR){
+            if (!StrCmp(misc->h->str,"dev") &&
+                  !StrCmp(cc->htc.fun->str,"Mem32DevIns")) {
+              Print("MOUNTIN GLOBAL fun=%s type=%X flags=%X gvflags=%X addr=%X "
+                    "data=%X data-rip=%X rip=%X offset=%X\\n",
+                    cc->htc.fun->str,misc->h->type,misc->flags,
+                    misc->h(CHashGlblVar*)->flags,misc->addr,
+                    misc->h(CHashGlblVar*)->data_addr,
+                    misc->h(CHashGlblVar*)->data_addr_rip,cc->aotc->rip,
+                    misc->offset);
+            }
             if (Bt(&cc->opts,OPTf_RAW_AOT) && misc->h->type&HTF_EXPORT) {
