--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -6,6 +6,7 @@
 #ifdef IMPORT_AIWNIOS_SYMS
 import U0 __HC_CmpCtrl_SetRawAOT(U8 *);
 import U0 __HC_CmpCtrl_PreserveSegments(U8 *);
+import U0 __HC_CmpCtrl_SetInterrupt(U8 *);
 import U0 __HC_SetAOTRelocBeforeRIP(U8 *,I64); //See aiwn.h(TAKES number of bytes before current RIP the symbol is located at)
 import U8 *__HC_CmpCtrlDel(U8 *);
 import U8 *__HC_ICAdd_Sqr(U8 *);
@@ -181,6 +182,7 @@
 extern U8 * __HC_ICAdd_OrEq(U8*);
 extern U8 * __HC_ICAdd_XorEq(U8*);
 extern U8 * __HC_ICAdd_ModEq(U8*);
+extern U0 __HC_CmpCtrl_SetInterrupt(U8 *);
 extern U8 * __HC_ICAdd_FReg(U8*,I64);
 extern U8 * __HC_ICAdd_Ds(U8 *,I64);
 extern U8 * __HC_ICAdd_IReg(U8*,I64 r,I64 rt,I64 ptrs);
@@ -1354,6 +1356,8 @@
   Bool force_segment=FALSE;
   I64 statics_sz=0;
   static I64 static_cnt=0;
+  if (cc->htc.fun && cc->htc.fun->flags&FSF_INTERRUPT)
+    __HC_CmpCtrl_SetInterrupt(acc);
   stat_nam=MStrPrint(".STATIC_%d",static_cnt++);
 //Turn all IC_IMM_I64s that are used as F64's to IC_IMM_I64
   for(cur=head->next;cur!=head;cur=cur->next) {
--- a/c/aiwn_lexparser.h
+++ b/c/aiwn_lexparser.h
@@ -187,6 +187,7 @@
 #define CCF_ICMOV_NO_USE_RAX         0x8
 #define CCF_RAW_AOT                  0x10
 #define CCF_PRESERVE_SEGMENTS        0x20
+#define CCF_INTERRUPT                0x40
   int64_t flags;
   CHashFun *cur_fun;
   CCodeCtrl *code_ctrl;
@@ -780,6 +781,7 @@
 void __HC_CmpCtrl_SetAOT(CCmpCtrl *cc);
 void __HC_CmpCtrl_SetRawAOT(CCmpCtrl *cc);
 void __HC_CmpCtrl_PreserveSegments(CCmpCtrl *cc);
+void __HC_CmpCtrl_SetInterrupt(CCmpCtrl *cc);
 
 CRPN *__HC_ICAdd_GetVargsPtr(CCodeCtrl *cc);
 
--- a/c/main.c
+++ b/c/main.c
@@ -928,6 +928,11 @@
   return 0;
 }
 
+static int64_t STK___HC_CmpCtrl_SetInterrupt(int64_t *stk) {
+  __HC_CmpCtrl_SetInterrupt((CCmpCtrl *)stk[0]);
+  return 0;
+}
+
 static int64_t STK__HC_ICAdd_RawBytes(int64_t *stk) {
   return (int64_t)__HC_ICAdd_RawBytes((CCmpCtrl *)stk[0], (char *)stk[1],
                                       stk[2]);
@@ -1756,6 +1761,8 @@
     PrsAddSymbol("__HC_CmpCtrl_SetRawAOT", STK___HC_CmpCtrl_SetRawAOT, 1);
     PrsAddSymbol("__HC_CmpCtrl_PreserveSegments",
                  STK___HC_CmpCtrl_PreserveSegments, 1);
+    PrsAddSymbol("__HC_CmpCtrl_SetInterrupt", STK___HC_CmpCtrl_SetInterrupt,
+                 1);
     PrsAddSymbol("__HC_ICAdd_Typecast", STK___HC_ICAdd_Typecast, 3);
     PrsAddSymbol("__HC_ICAdd_SubCall", STK___HC_ICAdd_SubCall, 2);
     PrsAddSymbol("__HC_ICAdd_SubProlog", STK___HC_ICAdd_SubProlog, 1);
--- a/c/parser.c
+++ b/c/parser.c
@@ -5158,6 +5158,10 @@
 void __HC_CmpCtrl_PreserveSegments(CCmpCtrl *cc) {
   cc->flags |= CCF_PRESERVE_SEGMENTS;
 }
+
+void __HC_CmpCtrl_SetInterrupt(CCmpCtrl *cc) {
+  cc->flags |= CCF_INTERRUPT;
+}
 HC_IC_BINDING(HC_ICAdd_ToBool, IC_TO_BOOL);
 HC_IC_BINDING(HC_ICAdd_BT, IC_BT);
 HC_IC_BINDING(HC_ICAdd_BTC, IC_BTC);
--- a/c/x86_64_backend.c
+++ b/c/x86_64_backend.c
@@ -387,6 +387,16 @@
   return len;
 }
 
+int64_t X86Iretq(char *to, int64_t unused) {
+  int64_t len = 0;
+  char buf[16];
+  if (!to)
+    to = buf;
+  ADD_U8(0x48);
+  ADD_U8(0xcf);
+  return len;
+}
+
 int64_t X86MovImm(char *to, int64_t a, int64_t off) {
   int64_t len = 0;
   char buf[16];
@@ -5804,7 +5814,9 @@
 
   off = -old_regs_start;
   AIWNIOS_ADD_CODE(X86Leave, 0);
-  if (is_vargs) {
+  if (cctrl->flags & CCF_INTERRUPT) {
+    AIWNIOS_ADD_CODE(X86Iretq, 0);
+  } else if (is_vargs) {
     AIWNIOS_ADD_CODE(X86Ret, 0);
   } else {
     AIWNIOS_ADD_CODE(X86Ret, 8 * arg_cnt);
