--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -361,5 +361,9 @@ Bool X86OpcodeUsesRel8(CX86Opcode *opc)
   return FALSE;
 }
+Bool X86OpcodeUsesSignExtendedImm8(CX86Opcode *opc)
+{
+  return opc->po==0x83 || opc->po==0x6B || opc->po==0x6A;
+}
 Bool ModeMatch(U16 md,CX86AddrMode *mode,I64 mem_sz,Bool mem_only=FALSE) {
   switch(md.u8[0]) {
     case X86_OPERAND_R8:
@@ -676,7 +681,11 @@ Bool OpcodePassesModes(CX86Opcode *opc,CX86AddrMode **modes) {
       if(opc->operands[idx].u8[0]==X86_OPERAND_EXT||
             opc->operands[idx].u8[0]==X86_OPERAND_SAME_RMx) goto nxt_arg;
       if(a>=opc->arity) break;
-      if(ModeMatch(opc->operands[idx],modes[a],mem_sz)) {
+      if(ModeMatch(opc->operands[idx],modes[a],mem_sz) &&
+            !(opc->operands[idx].u8[0]==X86_OPERAND_IMM8 &&
+              X86OpcodeUsesSignExtendedImm8(opc) &&
+              modes[a]->type==X86MT_IMM &&
+              !(I8_MIN<=modes[a]->imm<=I8_MAX))) {
 	a++;
 	goto nxt_arg;
       }
