diff --git a/Src/AsmX64.HC b/Src/AsmX64.HC
index 251a4f1..a68c5e4 100644
--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -106,6 +106,7 @@ CX86AddrMode *PrsX86AddrMode(CCmpCtrl *cc) {
   CX86AddrMode *mode=CAlloc(sizeof CX86AddrMode);
   CHashReg *rg=NULL;
   CHashClass *tmpc=NULL;
+  CHashGeneric *asm_kw;
   CMemberLst *tmpm;
   U8 *lookahead;
   I64 (*machine_code)(),cur_i64;
@@ -117,6 +118,31 @@ CX86AddrMode *PrsX86AddrMode(CCmpCtrl *cc) {
   mode->label=NULL;
 prefix:
+  asm_kw=NULL;
+  if(cc->token==TK_IDENT&&cc->hash_entry&&
+        cc->hash_entry->type&(HTT_CLASS|HTT_INTERNAL_TYPE)) {
+    lookahead=cc->cur_buf_ptr;
+    while(*lookahead&&*lookahead<=' ') lookahead++;
+    tmpc=cc->hash_entry;
+    if(!(tmpc->type&HTT_CLASS&&tmpc->member_lst_and_root&&
+          (cc->flags&CCF_USE_LAST_U16&&cc->last_U16=='.'||*lookahead=='.'))) {
+      mode->operand_sz=tmpc->size;
+      Lex(cc);
+      if(cc->token!='[')
+        goto size_prefix_consumed;
+    }
+  }
+  if(cc->hash_entry&&
+        cc->hash_entry->type&HTG_TYPE_MASK==HTT_ASM_KEYWORD)
+    asm_kw=cc->hash_entry;
+  else if(cc->token==TK_IDENT)
+    asm_kw=HashFind(cc->cur_str,cmp.asm_hash,HTT_ASM_KEYWORD);
+  if(asm_kw&&AKW_I8<=asm_kw->user_data0<=AKW_U64) {
+    mode->operand_sz=1<<((asm_kw->user_data0-AKW_I8)>>1);
+    Lex(cc);
+    if(cc->token!='[')
+      goto size_prefix_consumed;
+  }
   if(cc->token==TK_IDENT) {
     tmpc=HashFind(cc->cur_str,cmp.asm_hash,HTT_INTERNAL_TYPE);
     if(tmpc) {
       mode->operand_sz=tmpc->size;
