diff --git a/Src/AsmX64.HC b/Src/AsmX64.HC
index 07e50c51..a2cd073f 100644
--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -1575,7 +1575,7 @@ CAiwniosAOTBlob *PrsAsmBlk(CCmpCtrl *cc,I64 flags=0) {
   I64 old_flags=cc->flags;
   I64 old_class_dol_offset=cc->class_dol_offset;
   CHashGeneric *asm_entry;
-  Bool one_ins=FALSE,has_zero_arity,has_unresolved;
+  Bool one_ins=FALSE,has_zero_arity,has_unresolved,wants_rexw;
   if(flags&CMPF_ONE_ASM_INS) {
     one_ins=TRUE;
     goto enter0;
@@ -1653,15 +1653,24 @@ enter:
       if (modes[tmp]->machine_code)
         has_unresolved=TRUE;
     best=NULL,best_sz=I64_MAX;
+    wants_rexw=FALSE;
+    if (!arg_cnt && seg_size==64)
+      for(opc=first_opc;opc;opc=opc->next)
+        if(!StrCmp(first_opc->str,opc->str) &&
+              opc->arity==arg_cnt && OpcodePassesModes(opc,modes) &&
+              opc->flags&X86_OPCF_REXW)
+          wants_rexw=TRUE;
     for(opc=first_opc;opc;opc=opc->next) {
       if(!StrCmp(first_opc->str,opc->str))
        if(opc->arity==arg_cnt && OpcodePassesModes(opc,modes)) {
-	  sz=X86EncodeOpcode(len,seg_size,opc,modes,NULL,NULL);
-	  if (has_unresolved && X86OpcodeUsesRel8(opc))
-	    sz+=0x10000;
-	  if(sz<best_sz) {
-	    best_sz=sz;
-	    best=opc;
+	  if(!wants_rexw || opc->flags&X86_OPCF_REXW) {
+	    sz=X86EncodeOpcode(len,seg_size,opc,modes,NULL,NULL);
+	    if (has_unresolved && X86OpcodeUsesRel8(opc))
+	      sz+=0x10000;
+	    if(sz<best_sz) {
+	      best_sz=sz;
+	      best=opc;
+	    }
 	  }
         }
     }
