--- a/Src/KernelA.HH
+++ b/Src/KernelA.HH
@@ -2216,4 +2216,5 @@
   I64	aot_depth,pmt_line;
 //Added by nroot,used for @@labels
   I64 asm_local_scope;
+  I64 raw_aot_size;
 };
--- a/Src/PrsStmt.HC
+++ b/Src/PrsStmt.HC
@@ -1043,4 +1043,6 @@
 		    if (tmpaot=PrsAsmBlk(cc)) {
+		      if (!cc->htc.fun && Bt(&cc->opts,OPTf_RAW_AOT))
+			cc->raw_aot_size=tmpaot(CAiwniosAOTBlob*)->len;
 		      //If we are not in a function,we will jump past the assembly block
 		      // to avoid running it
 		      if(!cc->htc.fun && !Bt(&cc->opts,OPTf_RAW_AOT)) {
@@ -1057,4 +1059,6 @@
 		    if (tmpaot=PrsAsmBlk(cc)) {
+		      if (!cc->htc.fun && Bt(&cc->opts,OPTf_RAW_AOT))
+			cc->raw_aot_size=tmpaot(CAiwniosAOTBlob*)->len;
 		      //If we are not in a function,we will jump past the assembly block
 		      // to avoid running it
 		      if(!cc->htc.fun && !Bt(&cc->opts,OPTf_RAW_AOT)) {
@@ -1071,4 +1075,6 @@
 		    if (tmpaot=PrsAsmBlk(cc)) {
+		      if (!cc->htc.fun && Bt(&cc->opts,OPTf_RAW_AOT))
+			cc->raw_aot_size=tmpaot(CAiwniosAOTBlob*)->len;
 		      //If we are not in a function,we will jump past the assembly block
 		      // to avoid running it
 		      if(!cc->htc.fun && !Bt(&cc->opts,OPTf_RAW_AOT)) {
--- a/Src/CMain.HC
+++ b/Src/CMain.HC
@@ -5,5 +5,6 @@
   CRPN *rpn;
   LBtr(&cc->flags,CCf_PASS_TRACE_PRESENT);
+  cc->raw_aot_size=0;
   if (cc->aot_depth==2)
     COCPush(cc);
   COCInit(cc);
@@ -77,4 +78,5 @@
       }
-      if(cc->flags&CCF_AOT_COMPILE)
+      if(cc->flags&CCF_AOT_COMPILE &&
+            !cc->raw_aot_size)
 //ARM likes 4byte align for instrcutions
         while(cc->aotc->rip%4)
@@ -87,7 +89,12 @@
         if (cc->flags&CCF_AOT_COMPILE) {
 	  j=cc->aotc->rip;
-	  k=(size+7)>>3;
-	  for (i=0;i<k;i++)
-	    AOTStoreCodeU64(cc,res[i]);
+          if (cc->raw_aot_size)
+            for (i=0;i<cc->raw_aot_size;i++)
+              AOTStoreCodeU8(cc,res(U8*)[i]);
+          else {
+            k=(size+7)>>3;
+            for (i=0;i<k;i++)
+              AOTStoreCodeU64(cc,res[i]);
+          }
 	  Free(res);
 	  res=j;
