--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -1321,5 +1321,6 @@ U8 *AiwniosCompile(CCmpCtrl *cc,I64 *res_sz=NULL,CDbgInfo **info,Bool byte_code
   CRPN *head=&cc->coc.coc_head,*cur,*new;
   CCodeMisc *misc,**table;
+  Bool relocate_glbl;
   CMemberLst *mlst,*mlst2;
   COptMemberVar *mv=NULL;
   CHashImport *h;
@@ -2230,2 +2231,23 @@ U8 *AiwniosCompile(CCmpCtrl *cc,I64 *res_sz=NULL,CDbgInfo **info,Bool byte_code
-	    if(!Globalify(misc->addr,&fill_user_data))
-		  goto norm_reloc;
+            if (!StrCmp(misc->h->str,"dev") &&
+                  !StrCmp(cc->htc.fun->str,"Mem32DevIns"))
+              Print("MOUNTIN DEV type=%X flags=%X data-rip=%X offset=%X\\n",
+                    misc->h->type,misc->h(CHashGlblVar*)->flags,
+                    misc->h(CHashGlblVar*)->data_addr_rip,misc->offset);
+            if (!Globalify(misc->addr,&fill_user_data)) {
+              relocate_glbl=FALSE;
+              if (!(misc->h(CHashGlblVar*)->flags&GVF_EXTERN))
+                relocate_glbl=TRUE;
+              if (misc->h(CHashGlblVar*)->flags&GVF_ALIAS)
+                relocate_glbl=TRUE;
+              if (Bt(&cc->opts,OPTf_RAW_AOT) && relocate_glbl) {
+                foff=misc->h(CHashGlblVar*)->data_addr_rip+misc->offset;
+                WriteProtectMemCpy(misc->addr,&foff,8);
+                tmpabs=CAlloc(sizeof(CAOTAbsAddr));
+                tmpabs->next=cc->aotc->abss;
+                tmpabs->type=AAT_ADD_U64;
+                tmpabs->rip=cc->aotc->rip+
+                      (misc->addr(U8*)-machine_code);
+                cc->aotc->abss=tmpabs;
+              } else
+                goto norm_reloc;
+            }
