--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -825,9 +825,29 @@ I64 X86EncodeOpcode(I64 rip,I64 seg_size,CX86Opcode *opc,CX86AddrMode **modes,U
   I64 i,ext=-1,a;
   I64 operand_sz=-1,len=0,sz,indir_mode_idx=-1;
   I64 mask=(1<<opc->arity)-1;
-  Bool is_x_sz=FALSE,wants_modrm_sib=FALSE;
+  Bool fixed_opsize16=FALSE,fixed_opsize32=FALSE,has_wide_reg=FALSE;
+  Bool is_x_sz=FALSE,wants_modrm_sib=FALSE;
   I64 rex=0,rel_sz=0; //Upper 4 bits will be 0x4 if needed
   if(!to) to=dummy;
+  for(i=0;i!=4;i++)
+    switch(opc->operands[i].u8[0]) {
+      case X86_OPERAND_CONST_R16:
+      case X86_OPERAND_IMM16:
+      case X86_OPERAND_M16:
+      case X86_OPERAND_R16:
+      case X86_OPERAND_RM16:
+        fixed_opsize16=TRUE;
+        break;
+      case X86_OPERAND_IMM32:
+      case X86_OPERAND_M32:
+      case X86_OPERAND_R32:
+      case X86_OPERAND_RM32:
+        fixed_opsize32=TRUE;
+    }
+  for(a=0;a!=opc->arity;a++)
+    if(modes[a]->type==X86MT_REG &&
+          (modes[a]->reg_type==REGT_R32||modes[a]->reg_type==REGT_R64))
+      has_wide_reg=TRUE;
   CX86AddrMode *indir_mode=NULL;
   CX86AddrMode *reg_mode=NULL;
   CAsmPatch *patch;
@@ -1069,17 +1089,18 @@ I64 X86EncodeOpcode(I64 rip,I64 seg_size,CX86Opcode *opc,CX86AddrMode **modes,U
   if(segment_override!=-1)
     to[len++]=segment_override;
 
-  if(opc->pf!=-1) {
+  if(opc->pf!=-1 && !(opc->pf==0x66&&seg_size==16&&fixed_opsize16)) {
     to[len++]=opc->pf;  
   }
 //Prefix group 3(Operand size prefix)
   if(is_x_sz&&(seg_size==16&&operand_sz==4||seg_size!=16&&operand_sz==2)||
         opc->flags&X86_OPCF_OPSIZE16&&seg_size!=16||
-        opc->flags&X86_OPCF_OPSIZE32&&seg_size==16) {
+        (opc->flags&X86_OPCF_OPSIZE32||fixed_opsize32&&has_wide_reg)&&
+          seg_size==16) {
     to[len++]=0x66;
   }
 //Prefix group 4(Address size ovveride)
-  if (seg_size==16&&wants_modrm_sib ||
+  if (seg_size==16&&indir_mode ||
         seg_size==64&&indir_mode&&indir_mode->short_addr==1)
     to[len++]=0x67;
 
