--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -96,6 +96,7 @@
 import U8 *__HC_ICAdd_Bsr(U8*);
 import U8 *__HC_ICAdd_Rdtsc(U8*);
 import U8 *__HC_ICAdd_QueInit(U8*);
+import U8 *__HC_ICAdd_InvlPg(U8*);
 import U8 *__HC_ICAdd_Ret(U8*);
 import U8 *__HC_ICAdd_Arg(U8*,I64);
 import U8 *__HC_ICAdd_SetFrameSize(U8*,I64);
@@ -202,6 +203,7 @@
 extern U8 *__HC_ICAdd_Bsr(U8*);
 extern U8 *__HC_ICAdd_Rdtsc(U8*);
 extern U8 *__HC_ICAdd_QueInit(U8*);
+extern U8 *__HC_ICAdd_InvlPg(U8*);
 extern U8 *__HC_ICAdd_Ret(U8*);
 extern U8 *__HC_ICAdd_Arg(U8*,I64);
 extern U8 *__HC_ICAdd_SetFrameSize(U8*,I64);
@@ -1988,6 +1990,8 @@
     new=__HC_ICAdd_Rdtsc(cc2);
     break;case IC_QUE_INIT:
     new=__HC_ICAdd_QueInit(cc2);
+    break;case IC_INVLPG:
+    new=__HC_ICAdd_InvlPg(cc2);
     break;case IC_SQRT:
     new=__HC_ICAdd_Sqrt(cc2);
     break;case IC_DEREF:
--- a/c/aiwn_lexparser.h
+++ b/c/aiwn_lexparser.h
@@ -243,6 +243,7 @@
   IC_BSR,
   IC_RDTSC,
   IC_QUE_INIT,
+  IC_INVLPG,
   IC_POS,
   IC_NAME,
   IC_STR, //->code_misc
@@ -690,6 +691,7 @@
 HC_IC_BINDINGH(HC_ICAdd_Bsr)
 HC_IC_BINDINGH(HC_ICAdd_Rdtsc)
 HC_IC_BINDINGH(HC_ICAdd_QueInit)
+HC_IC_BINDINGH(HC_ICAdd_InvlPg)
 HC_IC_BINDINGH(HC_ICAdd_And)
 HC_IC_BINDINGH(HC_ICAdd_EqEq)
 HC_IC_BINDINGH(HC_ICAdd_Ret)
--- a/c/main.c
+++ b/c/main.c
@@ -1207,6 +1207,9 @@
 static int64_t STK___HC_ICAdd_QueInit(int64_t *stk) {
   return __HC_ICAdd_QueInit(stk[0]);
 }
+static int64_t STK___HC_ICAdd_InvlPg(int64_t *stk) {
+  return __HC_ICAdd_InvlPg(stk[0]);
+}
 static int64_t STK___HC_ICAdd_Ret(int64_t *stk) {
   return __HC_ICAdd_Ret(stk[0]);
 }
@@ -1824,6 +1827,7 @@
     PrsAddSymbol("__HC_ICAdd_Bsr", STK___HC_ICAdd_Bsr, 1);
     PrsAddSymbol("__HC_ICAdd_Rdtsc", STK___HC_ICAdd_Rdtsc, 1);
     PrsAddSymbol("__HC_ICAdd_QueInit", STK___HC_ICAdd_QueInit, 1);
+    PrsAddSymbol("__HC_ICAdd_InvlPg", STK___HC_ICAdd_InvlPg, 1);
     PrsAddSymbol("__HC_ICAdd_Ret", STK___HC_ICAdd_Ret, 1);
     PrsAddSymbol("__HC_ICAdd_Arg", STK___HC_ICAdd_Arg, 2);
     PrsAddSymbol("__HC_ICAdd_SetFrameSize", STK___HC_ICAdd_SetFrameSize, 2);
--- a/c/parser.c
+++ b/c/parser.c
@@ -244,6 +244,7 @@
     break;
   case IC_TYPECAST:
   case IC_QUE_INIT:
+  case IC_INVLPG:
     goto unop;
     break;
   case IC_COMMA:
@@ -805,6 +806,9 @@
     printf("QUE_INIT\n");
     goto unop;
     break;
+  case IC_INVLPG:
+    printf("INVLPG\n");
+    goto unop;
   case IC_NAME:
     printf("NAME:%s\n", rpn->string);
     goto ret;
@@ -1334,6 +1338,9 @@
     len = AddBytes(to, len, "$PURPLE$QUE_INIT$FD$\n");
     goto unop;
     break;
+  case IC_INVLPG:
+    len = AddBytes(to, len, "$PURPLE$INVLPG$FD$\n");
+    goto unop;
   case IC_NAME:
     goto ret;
     break;
@@ -3488,6 +3495,10 @@
     AssignRawTypeToNode(ccmp, ICArgN(rpn, 0));
     return rpn->raw_type = RT_U0;
     break;
+  case IC_INVLPG:
+    AssignRawTypeToNode(ccmp, ICArgN(rpn, 0));
+    return rpn->raw_type = RT_U0;
+    break;
   case IC_BT:
   case IC_BTR:
   case IC_BTS:
@@ -4657,6 +4668,7 @@
 HC_IC_BINDING(HC_ICAdd_Bsr, IC_BSR);
 HC_IC_BINDING(HC_ICAdd_Rdtsc, IC_RDTSC);
 HC_IC_BINDING(HC_ICAdd_QueInit, IC_QUE_INIT);
+HC_IC_BINDING(HC_ICAdd_InvlPg, IC_INVLPG);
 HC_IC_BINDING(HC_ICAdd_And, IC_AND);
 HC_IC_BINDING(HC_ICAdd_Ret, IC_RET);
 
--- a/c/x86_64_backend.c
+++ b/c/x86_64_backend.c
@@ -1369,5 +1369,18 @@
   return len;
 }
 
+static int64_t X86InvlPg(char *to, int64_t scale, int64_t index,
+                         int64_t base, int64_t off) {
+  int64_t len = 0;
+  char buf[16];
+  if (!to)
+    to = buf;
+  SIB_BEGIN(1, 7, scale, index, base, off);
+  ADD_U8(0x0f);
+  ADD_U8(0x01);
+  SIB_END();
+  return len;
+}
+
 static int64_t X86IMul2SIB64(char *to, int64_t r, int64_t s, int64_t i,
                              int64_t b, int64_t off) {
@@ -2218,6 +2229,7 @@
   case IC_BSF:
   case IC_BSR:
   case IC_QUE_INIT:
+  case IC_INVLPG:
   case IC_SQRT:
   case IC_TO_BOOL:
   unop:
@@ -2898,6 +2910,7 @@
   case IC_BSF:
   case IC_BSR:
   case IC_QUE_INIT:
+  case IC_INVLPG:
     goto unop;
     break;
   case IC_POS:
@@ -6145,6 +6158,7 @@
   case IC_BSF:
   case IC_BSR:
   case IC_QUE_INIT:
+  case IC_INVLPG:
   case IC_LNOT:
   case IC_BNOT:
   case IC_POS:
@@ -6374,5 +6388,6 @@
       [IC_CURRENT_RIP] = &&ic_current_rip,
       [IC_RDTSC] = &&ic_rdtsc,
       [IC_QUE_INIT] = &&ic_que_init,
+      [IC_INVLPG] = &&ic_invlpg,
       [IC_RAW_BYTES] = &&ic_raw_bytes,
   };
@@ -6382,5 +6397,13 @@
   goto *poop_ants[rpn->type];
   do {
+  ic_invlpg:
+    next = ICArgN(rpn, 0);
+    next->res.keep_in_tmp = 1;
+    code_off = __OptPassFinal(cctrl, next, bin, code_off);
+    code_off = PutICArgIntoReg(cctrl, &next->res, RT_PTR, RAX, bin, code_off);
+    AIWNIOS_ADD_CODE(X86InvlPg, -1, -1, MIR(cctrl, next->res.reg), 0);
+    break;
+
   ic_que_init:
     next = ICArgN(rpn, 0);
     next->res.keep_in_tmp = 1;
--- a/c/optpass.c
+++ b/c/optpass.c
@@ -702,6 +702,7 @@
       break;
     case IC_DEREF:
     case IC_QUE_INIT:
+    case IC_INVLPG:
       goto next1;
       break;
     case IC_AND:
@@ -1130,6 +1131,7 @@
       break;
     case IC_DEREF:
     case IC_QUE_INIT:
+    case IC_INVLPG:
       continue;
       break;
     case IC_AND:
