--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -658,7 +658,8 @@
   for(mem_sz=2;mem_sz<=8;mem_sz*=2) {
     a=0;
     for(idx=0;opc->operands[idx].u8[0];idx++) {
-      if(opc->operands[idx].u8[0]==X86_OPERAND_EXT) goto nxt_arg;
+      if(opc->operands[idx].u8[0]==X86_OPERAND_EXT||
+            opc->operands[idx].u8[0]==X86_OPERAND_SAME_RMx) goto nxt_arg;
       if(a>=opc->arity) break;
       if(ModeMatch(opc->operands[idx],modes[a],mem_sz)) {
 	a++;
@@ -945,7 +946,15 @@
 	      a++;
 	      break;
 	    }
-	  case X86_OPERAND_R8:
+	  case X86_OPERAND_SAME_RMx:
+	    if(!indir_mode) {
+	      if(!a)
+	        throw('AsmItrn');
+	      indir_mode_idx=a-1;
+	      indir_mode=modes[a-1];
+	    }
+	    break;
+	  case X86_OPERAND_R8:
 	  case X86_OPERAND_R16:
 	  case X86_OPERAND_R32:
 	  case X86_OPERAND_R64:
@@ -1162,7 +1171,8 @@
 		  len+=8;
 	      }
 	    }
-	    if(opc->operands[i].u8[0]!=X86_OPERAND_EXT)
+	    if(opc->operands[i].u8[0]!=X86_OPERAND_EXT&&
+	          opc->operands[i].u8[0]!=X86_OPERAND_SAME_RMx)
 	      a++;
 	  }
         } else
--- a/Src/UAsmX64.HC
+++ b/Src/UAsmX64.HC
@@ -59,6 +59,8 @@
 //TODO,Nroot is not implementeing vector types now.
 #define X86_OPERAND_M128 49
 #define X86_OPERAND_CONST_SREG 50
+//The ModR/M source is the preceding explicit register operand.
+#define X86_OPERAND_SAME_RMx 51
 
 class CX86Opcode:CHash {
   //-1 for absent
@@ -111,7 +113,8 @@
     if(idx>argc) {
       throw('UAsmInit');
     }
-    if(argv[idx]!=X86_OPERAND_EXT)
+    if(argv[idx]!=X86_OPERAND_EXT&&
+          argv[idx]!=X86_OPERAND_SAME_RMx)
       opc->arity++;
     switch(argv[idx]) {
       start:
@@ -1356,6 +1359,8 @@
   AO("IMUL",-1,-1,0xf6,-1,X86_OPCF_MODRM,X86_OPERAND_RM8,X86_OPERAND_EXT,5);
   AO("IMUL",-1,-1,0xf7,-1,X86_OPCF_MODRM,X86_OPERAND_RMx,X86_OPERAND_EXT,5);
   AO("IMUL2",-1,0xf,0xaf,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_RMx);
+  AO("IMUL2",-1,-1,0x69,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_SAME_RMx,X86_OPERAND_IMMx);
+  AO("IMUL2",-1,-1,0x6b,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_SAME_RMx,X86_OPERAND_IMM8);
 
   AO("IN",-1,-1,0xe4,-1,0,X86_OPERAND_CONST_R8,REG_RAX,X86_OPERAND_IMM8);
   AO("IN",-1,-1,0xe5,-1,0,X86_OPERAND_CONST_R32,REG_RAX,X86_OPERAND_IMM8);
