--- a/Kernel/KLoad.HC
+++ b/Kernel/KLoad.HC
@@ -67,7 +67,7 @@
 U0 LoadPass1(U8 *src,U8 *module_base,I64 ld_flags)
 {
   U8 *ptr2,*ptr3,*st_ptr;
-  I64 i,j,cnt,etype;
+  I64 i,j,k,cnt,etype;
   CHashExport *tmpex=NULL;
   while (etype=*src++) {
     i=*src(U32 *)++;
@@ -128,7 +128,10 @@
 
       start:
 	case IET_DATA_HEAP:
-	  ptr3=MAlloc(*src(I64 *)++);
+	  k=*src(I64 *)++;
+	  ptr3=MAlloc(k);
+	  MemCpy(ptr3,src,k);
+	  src+=k;
 	  break;
 	case IET_ZEROED_DATA_HEAP:
 	  ptr3=CAlloc(*src(I64 *)++);
@@ -144,7 +147,10 @@
 	cnt=i;
 	for (j=0;j<cnt;j++) {
 	  ptr2=module_base+*src(U32 *)++;
-	  *ptr2(I64 *)+=ptr3;
+	  if (etype==IET_DATA_HEAP)
+	    *ptr2(I64 *)+=ptr3+*src(U32 *)++;
+	  else
+	    *ptr2(I64 *)+=ptr3;
 	}
 	break;
     }
@@ -154,7 +160,7 @@
 U0 LoadPass2(U8 *src,U8 *module_base,I64)
 {
   U8 *st_ptr;
-  I64 i,etype;
+  I64 i,j,etype;
   while (etype=*src++) {
     i=*src(U32 *)++;
     st_ptr=src;
@@ -171,6 +177,9 @@
 	src+=4+sizeof(U32)*i;
 	break;
       case IET_DATA_HEAP:
+	j=*src(I64 *)++;
+	src+=j+sizeof(U32)*2*i;
+	break;
       case IET_ZEROED_DATA_HEAP:
 	src+=8+sizeof(U32)*i;
 	break;
