--- a/Kernel/KStart32.HC
+++ b/Kernel/KStart32.HC
@@ -112,6 +112,5 @@
 	CALL	SYS_INIT_16MEG_SYS_CODE_BP
 	PUSH	U32 0	//Return from next call will be 64-bit
 	CALL	SYS_ENTER_LONG_MODE
-
-	//We fall-thou to $LK,"KStart64",A="FL:::/Kernel/KStart64.HC,1"$, next.
+	JMP	I32 CORE0_64BIT_INIT
 }
--- a/Kernel/KStart64.HC
+++ b/Kernel/KStart64.HC
@@ -1,5 +1,6 @@
 asm {
 USE64
+CORE0_64BIT_INIT::
 	BTS	U32 [SYS_RUN_LEVEL],RLf_64BIT
 
 	FNINIT
