diff --git a/Demo/Lectures/InterruptDemo.HC b/Demo/Lectures/InterruptDemo.HC
index 8048c52c..d094a23b 100644
--- a/Demo/Lectures/InterruptDemo.HC
+++ b/Demo/Lectures/InterruptDemo.HC
@@ -14,7 +14,7 @@ F_UNARY_MINUS_INT::
 	FCHS
 	FSTP	U64 [RSP]
 	POP	RAX
-	IRET
+	IRETQ
 
 //This does the same thing, but not as
 //an interrupt.
diff --git a/Demo/Lectures/Ring3.HC b/Demo/Lectures/Ring3.HC
index 25a5a134..0a37e1b7 100644
--- a/Demo/Lectures/Ring3.HC
+++ b/Demo/Lectures/Ring3.HC
@@ -21,7 +21,7 @@ INT_TO_RING0::	//Set to handle general protection 0xD fault temporarily.
 	PUSH	U32 CGDT.cs64
 	MOV	RAX,U64 [&new_rip]
 	PUSH	RAX
-	IRET
+	IRETQ
 }
 
 U0 Ring3Demo()
@@ -51,7 +51,7 @@ U0 Ring3Demo()
     PUSH	U32 CGDT.cs64_ring3+3
     LEA	RAX,[R3_START]
     PUSH	RAX
-    IRET
+	IRETQ
 
     R3_START:
     INC	U64 [SYS_PROGRESS1]
diff --git a/Kernel/KInts.HC b/Kernel/KInts.HC
index 00652181..73cf8d77 100644
--- a/Kernel/KInts.HC
+++ b/Kernel/KInts.HC
@@ -11,7 +11,7 @@ INT_WAKE::
 	MOV	U32 [RAX+RDX],0
 	POP	RAX
 	POP	RDX
-	IRET
+	IRETQ
 
 IRQ_TIMER::  //I_TIMER
 	CALL	TASK_CONTEXT_SAVE
diff --git a/Kernel/KStart64.HC b/Kernel/KStart64.HC
index ab887950..41d9329e 100644
--- a/Kernel/KStart64.HC
+++ b/Kernel/KStart64.HC
@@ -78,7 +78,7 @@ SYS_RAM_REBOOT:: //This gets copied high.  $LK,"SYS_RAM_REBOOT",A="FF:::/Adam/Op
 	PUSH	U32 CGDT.cs32
 	LEA	RAX,[@@10]
 	PUSH	RAX
-	IRET
+	IRETQ
 USE32
 @@10:	WBINVD
 
diff --git a/Kernel/PCIBIOS.HC b/Kernel/PCIBIOS.HC
index 37dd1996..44deafad 100644
--- a/Kernel/PCIBIOS.HC
+++ b/Kernel/PCIBIOS.HC
@@ -91,7 +91,7 @@ _FAR_CALL32::
 	PUSH	U32 CGDT.cs32
 	LEA	RAX,[@@15]
 	PUSH	RAX
-	IRET
+	IRETQ
 USE32
 @@15:
 	WBINVD
diff --git a/Kernel/Sched.HC b/Kernel/Sched.HC
index a86b99ac..cac540f7 100644
--- a/Kernel/Sched.HC
+++ b/Kernel/Sched.HC
@@ -142,7 +142,7 @@ _TASK_CONTEXT_RESTORE::
 	PUSH	CGDT.cs64
 	PUSH	U64 CTask.rip[RSI]
 	MOV	RSI,U64 CTask.rsi[RSI]
-	IRET
+	IRETQ
 //************************************
 END_RSI_TASK:
 	MOV	RAX,RSI
