--- a/Kernel/BlkDev/DskAddDev.HC
+++ b/Kernel/BlkDev/DskAddDev.HC
@@ -187,12 +187,7 @@
 {
   I64 ch;
   Bool probe;
-  #exe {
-    if (kernel_cfg->opts[CFG_DONT_PROBE])
-      StreamPrint("probe=FALSE;");
-    else
-      StreamPrint("probe=TRUE;");
-  };
+  probe=TRUE;
   if (!probe || !BootDVDProbeAll(bd)) {
     "\nDon't worry.  This is not a product\n"
 	  "registration.	TempleOS just needs the\n"
@@ -230,11 +225,5 @@
   blkdev.drvs=CAlloc(sizeof(CDrv)*DRVS_NUM);
   for (i=0;i<DRVS_NUM;i++)
     blkdev.let_to_drv[i]=&blkdev.drvs[i];
-  #exe {
-    if (kernel_cfg->opts[CFG_MOUNT_IDE_AUTO])
-      StreamPrint("MountIDEAuto;");
-    StreamPrint("#exe {Option(OPTf_WARN_PAREN,OFF);}");
-    StreamDoc(kernel_cfg->add_dev);
-    StreamPrint("#exe {Option(OPTf_WARN_PAREN,ON);}");
-  };
+  MountIDEAuto;
 }
--- a/Kernel/KStart16.HC
+++ b/Kernel/KStart16.HC
@@ -79,10 +79,7 @@
 
 	MOV	AX,0x4F02
 	MOV	BX,0x12		//640x480 16 color
-#exe {
-  if (!kernel_cfg->opts[CFG_TEXT_MODE])
-    StreamPrint("INT 0x10");	//Enable VGA
-};
+	INT	0x10		//Enable VGA
 	CMP	AX,0x004F
 	JNE	@@10		//Jmp if fail
 	BTS	U32 [SYS_RUN_LEVEL],RLf_VGA
--- a/Kernel/KStart32.HC
+++ b/Kernel/KStart32.HC
@@ -14,20 +14,13 @@
 MEM_PML3::		DU64	0;
 MEM_PML4::		DU64	0;
 
-#exe {
-  StreamPrint(
-"SYS_MEM_INIT_FLAG::	DU8	%d;"
-"SYS_MEM_INIT_VAL::	DU8	%d;"
-"SYS_HEAP_INIT_FLAG::	DU8	%d;"
-"SYS_HEAP_INIT_VAL::	DU8	%d;"
-"SYS_VAR_INIT_FLAG::	DU8	%d;"
-"SYS_VAR_INIT_VAL::	DU8	%d;"
-"SYS_STAFF_MODE_FLAG::	DU8	%d;",
-kernel_cfg->opts[CFG_MEM_INIT], kernel_cfg->mem_init_val,
-kernel_cfg->opts[CFG_HEAP_INIT],kernel_cfg->heap_init_val,
-kernel_cfg->opts[CFG_VAR_INIT], kernel_cfg->var_init_val,
-kernel_cfg->opts[CFG_STAFF_MODE]);
-};
+SYS_MEM_INIT_FLAG::	DU8	0;
+SYS_MEM_INIT_VAL::	DU8	0;
+SYS_HEAP_INIT_FLAG::	DU8	0;
+SYS_HEAP_INIT_VAL::	DU8	0;
+SYS_VAR_INIT_FLAG::	DU8	0;
+SYS_VAR_INIT_VAL::	DU8	0;
+SYS_STAFF_MODE_FLAG::	DU8	0;
 SYS_HEAP_DBG_FLAG::	DU8	_CFG_HEAP_DBG;
 
 	ALIGN	8,0
--- a/Kernel/KMain.HC
+++ b/Kernel/KMain.HC
@@ -13,16 +13,8 @@
   blkdev.dft_iso_c_filename	=AStrNew(DFT_ISO_C_FILENAME);
   blkdev.tmp_filename		=AStrNew("~/Tmp.DD.Z");
   blkdev.dvd_boot_is_good=TRUE;
-  #exe {
-    if (!kernel_cfg->mount_ide_auto_hd_let)
-      kernel_cfg->mount_ide_auto_hd_let='C';
-    if (!kernel_cfg->mount_ide_auto_cd_let)
-      kernel_cfg->mount_ide_auto_cd_let='T';
-    StreamPrint("blkdev.first_hd_drv_let=%d;",
-	  kernel_cfg->mount_ide_auto_hd_let);
-    StreamPrint("blkdev.first_dvd_drv_let=%d;",
-	  kernel_cfg->mount_ide_auto_cd_let);
-  }
+  blkdev.first_hd_drv_let='C';
+  blkdev.first_dvd_drv_let='T';
 
   DbgMode(ON);
   rev_bits_table=CAlloc(256);
@@ -37,12 +29,8 @@
   fp_getstr2=&SysGetStr2;
   KeyDevInit;
 
-  #exe {
-    StreamPrint("blkdev.boot_drv_let='%C';",kernel_cfg->boot_drv_let);
-    StreamPrint("#exe{Option(OPTf_WARN_PAREN,OFF);}");
-    StreamPrint("DskCacheInit(%s);",kernel_cfg->dsk_cache_size_exp);
-    StreamPrint("#exe{Option(OPTf_WARN_PAREN,ON);}");
-  };
+  blkdev.boot_drv_let='T';
+  DskCacheInit(Scale2Mem(2048,0x40000));
 
   pow10_I64=CAlloc(sizeof(F64)*(308+308+1));
   for (i=-308;i<309;i++)
@@ -175,12 +163,7 @@
   BlkDevsInitAll;
   "DskChg(':');\n";
   DskChg(':');
-  #exe {
-    StreamPrint("HomeSet(\"%s\");\n"
-	  "blkdev.ins_base0=%d;blkdev.ins_base1=%d;blkdev.ins_unit =%d;\n",
-	  kernel_cfg->home_dir,
-	  blkdev.ins_base0,blkdev.ins_base1,blkdev.ins_unit);
-  }
+  HomeSet("::/Home");
   Gs->idle_task->cur_dv=blkdev.let_to_drv[*blkdev.home_dir-'A'];
   DrvRep;
   if (blkdev.dvd_boot_is_good) {
@@ -193,21 +176,18 @@
   }
   LBts(&sys_run_level,RLf_BLKDEV);
 
-  #exe {
-    if (!kernel_cfg->opts[CFG_NO_MP])
-      StreamPrint("\"MultiCore Start\\n\\n\";"
-	    "Core0StartMP;"
-	    "LBts(&sys_run_level,RLf_MP);");
-  };
+  "MultiCore Start\n\n";
+  Core0StartMP;
+  LBts(&sys_run_level,RLf_MP);
 
   KbdMsInit;
   MsInit;
   KbdInit;
-  Spawn(&MsHardDrvrInstall);
 
   "Loading Compiler\n";
   Cd("/Compiler");
   Load("Compiler",LDF_SILENT);
   LBts(&sys_run_level,RLf_COMPILER);
+  Spawn(&MsHardDrvrInstall);
 
   DbgMode(OFF);
@@ -228,11 +208,4 @@
 
 asm {	ALIGN	16,OC_NOP
 SYS_KERNEL_END::
-#exe {
-  if (kernel_cfg->opts[CFG_DBG_DISTRO])
-    StreamPrint("DU8 0x%X-(SYS_KERNEL_END-SYS_KERNEL+"
-	"BOOT_RAM_BASE+sizeof(CBinFile)) DUP (0);"
-	"BINFILE \"%s\";",kernel_cfg->dbg_distro_start,
-			  kernel_cfg->dbg_distro_file);
-};
 }
--- a/Kernel/KEnd.HC
+++ b/Kernel/KEnd.HC
@@ -13,16 +13,8 @@
   blkdev.dft_iso_c_filename	=AStrNew(DFT_ISO_C_FILENAME);
   blkdev.tmp_filename		=AStrNew("~/Tmp.DD.Z");
   blkdev.dvd_boot_is_good=TRUE;
-  #exe {
-    if (!kernel_cfg->mount_ide_auto_hd_let)
-      kernel_cfg->mount_ide_auto_hd_let='C';
-    if (!kernel_cfg->mount_ide_auto_cd_let)
-      kernel_cfg->mount_ide_auto_cd_let='T';
-    StreamPrint("blkdev.first_hd_drv_let=%d;",
-	  kernel_cfg->mount_ide_auto_hd_let);
-    StreamPrint("blkdev.first_dvd_drv_let=%d;",
-	  kernel_cfg->mount_ide_auto_cd_let);
-  }
+  blkdev.first_hd_drv_let='C';
+  blkdev.first_dvd_drv_let='T';
 
   DbgMode(ON);
   rev_bits_table=CAlloc(256);
@@ -37,12 +29,8 @@
   fp_getstr2=&SysGetStr2;
   KeyDevInit;
 
-  #exe {
-    StreamPrint("blkdev.boot_drv_let='%C';",kernel_cfg->boot_drv_let);
-    StreamPrint("#exe{Option(OPTf_WARN_PAREN,OFF);}");
-    StreamPrint("DskCacheInit(%s);",kernel_cfg->dsk_cache_size_exp);
-    StreamPrint("#exe{Option(OPTf_WARN_PAREN,ON);}");
-  };
+  blkdev.boot_drv_let='T';
+  DskCacheInit(Scale2Mem(2048,0x40000));
 
   pow10_I64=CAlloc(sizeof(F64)*(308+308+1));
   for (i=-308;i<309;i++)
@@ -174,12 +162,7 @@
   BlkDevsInitAll;
   "DskChg(':');\n";
   DskChg(':');
-  #exe {
-    StreamPrint("HomeSet(\"%s\");\n"
-	  "blkdev.ins_base0=%d;blkdev.ins_base1=%d;blkdev.ins_unit =%d;\n",
-	  kernel_cfg->home_dir,
-	  blkdev.ins_base0,blkdev.ins_base1,blkdev.ins_unit);
-  }
+  HomeSet("::/Home");
   Gs->idle_task->cur_dv=blkdev.let_to_drv[*blkdev.home_dir-'A'];
   DrvRep;
   if (blkdev.dvd_boot_is_good) {
@@ -192,22 +175,19 @@
   }
   LBts(&sys_run_level,RLf_BLKDEV);
 
-  #exe {
-    if (!kernel_cfg->opts[CFG_NO_MP])
-      StreamPrint("\"MultiCore Start\\n\\n\";"
-	    "Core0StartMP;"
-	    "LBts(&sys_run_level,RLf_MP);");
-  };
+  "MultiCore Start\n\n";
+  Core0StartMP;
+  LBts(&sys_run_level,RLf_MP);
 
   KbdMsInit;
   MsInit;
   KbdInit;
-  Spawn(&MsHardDrvrInstall);
 
   "Loading Compiler\n";
   Cd("/Compiler");
   Load("Compiler",LDF_SILENT);
   LBts(&sys_run_level,RLf_COMPILER);
+  Spawn(&MsHardDrvrInstall);
 
   DbgMode(OFF);
   cnts.time_stamp_freq_initial=TimeCal;
@@ -227,11 +207,4 @@
 
 asm {	ALIGN	16,OC_NOP
 SYS_KERNEL_END::
-#exe {
-  if (kernel_cfg->opts[CFG_DBG_DISTRO])
-    StreamPrint("DU8 0x%X-(SYS_KERNEL_END-SYS_KERNEL+"
-	"BOOT_RAM_BASE+sizeof(CBinFile)) DUP (0);"
-	"BINFILE \"%s\";",kernel_cfg->dbg_distro_start,
-			  kernel_cfg->dbg_distro_file);
-};
 }
