--- a/Kernel/KernelB.HH
+++ b/Kernel/KernelB.HH
@@ -283,6 +283,8 @@ public _intern IC_CLFLUSH U0 CLFlush(U8 *a); //Flush cache line inst.
 public _intern IC_INVLPG
 	U0 InvlPg(U8 *a); //Invalidate page at addr inst.
 public _extern MEM_PAGE_SIZE U64 mem_page_size;
+public _extern MEM_PML2 U64 *mem_pml2;
+public _extern MEM_PML3 U64 *mem_pml3;
 
 #help_index "Task;Processor"
 public _intern IC_FS CTask *Fs(); //FS seg reg points to the current $LK,"CTask",A="MN:CTask"$.
--- a/Kernel/Mem/PageTables.HC
+++ b/Kernel/Mem/PageTables.HC
@@ -199,8 +199,8 @@ I64 *MemPageTable(U8 *a)
 {//Point to page table entry for addr.
   if (Bt(&mem_page_size,30))
-    return *MEM_PML3(U64 *)+a>>30*8;
+    return mem_pml3(U8 *)+(a>>30)*8;
   else
-    return *MEM_PML2(U64 *)+a>>21*8;
+    return mem_pml2(U8 *)+(a>>21)*8;
 }
