--- a/Src/PrsStmt.HC
+++ b/Src/PrsStmt.HC
@@ -967,6 +967,31 @@ U0 PrsTryBlk(CCmpCtrl *cc,I64 try_cnt)
   lb_done =COCMiscNew(cc,CMT_LABEL);
   lb_untry=COCMiscNew(cc,CMT_LABEL);
 
+  if (cc->flags&CCF_AOT_COMPILE && Bt(&cc->opts,OPTf_RAW_AOT)) {
+    RPNAddNew(cc,IC_ADDR_IMPORT,tmp_try,tmpc,FALSE);
+    RPNAddNew(cc,IC_GET_LABEL,lb_catch,tmpc,FALSE);
+    RPNAddNew(cc,IC_GET_LABEL,lb_untry,tmpc,FALSE);
+    RPNAddNew(cc,IC_CALL_INDIRECT,3,tmpc,FALSE);
+
+    PrsStmt(cc,try_cnt+1);
+
+    RPNAddNew(cc,IC_LABEL,lb_untry,0,FALSE);
+    RPNAddNew(cc,IC_ADDR_IMPORT,tmp_untry,tmpc,FALSE);
+    RPNAddNew(cc,IC_CALL_INDIRECT,1,tmpc,FALSE);
+    RPNAddNew(cc,IC_JMP,lb_done,0,FALSE);
+
+    if (PrsKeyWord(cc)!=KW_CATCH)
+      LexExcept(cc,"Missing 'catch' at");
+
+    Lex(cc);
+    RPNAddNew(cc,IC_LABEL,lb_catch,0,FALSE);
+    PrsStmt(cc,try_cnt+1);
+    RPNAddNew(cc,IC_IMM_I64,0,cmp.internal_types[RT_I64],FALSE);
+    RPNAddNew(cc,IC_RET,0,cc->htc.fun->return_class,FALSE);
+    RPNAddNew(cc,IC_LABEL,lb_done,0,FALSE);
+    return;
+  }
+
   //Changed by nroot
   RPNAddNew(cc,IC_ADDR_IMPORT,tmp_a_ent,tmpc,FALSE);
   RPNAddNew(cc,IC_ADDR_IMPORT,tmp_try,tmpc,FALSE);
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -87,5 +87,6 @@ import U8 *__HC_CodeMiscStrNew(U8 *,U8*,I64);
 import U8 *__HC_CodeMiscJmpTableNew(U8*,U8**,I64,I64);
 import U8 *__HC_ICAdd_Label(U8*,U8*);
+import U8 *__HC_ICAdd_LabelAddr(U8*,U8*);
 import U8 *__HC_ICAdd_Goto(U8*,U8*);
 import U8 *__HC_ICAdd_GotoIf(U8*,U8*)
 import U8 *__HC_ICAdd_Str(U8*,U8*); //Takes a AIWNIOS code misc
@@ -188,5 +189,6 @@ extern U8 *__HC_CodeMiscStrNew(U8 *,U8*,I64);
 extern U8 *__HC_CodeMiscJmpTableNew(U8*,U8**,I64,I64);
 extern U8 *__HC_ICAdd_Label(U8*,U8*);
+extern U8 *__HC_ICAdd_LabelAddr(U8*,U8*);
 extern U8 *__HC_ICAdd_Goto(U8*,U8*);
 extern U8 *__HC_ICAdd_GotoIf(U8*,U8*)
 extern U8 *__HC_ICAdd_Str(U8*,U8*); //Takes a AIWNIOS code misc
@@ -1818,5 +1820,10 @@ CRPN *CopyCode(CCmpCtrl *cc,CCodeCtrl *cc2,CAiwniosAOT *acc)
     break;case IC_LABEL:
       new=__HC_ICAdd_Label(cc2,cur->ic_data(CCodeMisc*)->addr);
+    break;case IC_GET_LABEL:
+      misc=cur->ic_data;
+      if (!misc->addr)
+        misc->addr=__HC_CodeMiscLabelNew(acc,&misc->addr2);
+      new=__HC_ICAdd_LabelAddr(cc2,misc->addr);
 break;case IC_ASM:
   __HC_ICAdd_Label(cc2,__HC_CodeMiscLabelNew(acc,&cur->ic_data(CAiwniosAOTBlob*)->final_start_addr));
    new=__HC_ICAdd_RawBytes(cc2,cur->ic_data(CAiwniosAOTBlob*)->data,cur->ic_data(CAiwniosAOTBlob*)->len);
--- a/Src/AIWNIOS_PrsExp.HC
+++ b/Src/AIWNIOS_PrsExp.HC
@@ -95,2 +95,1 @@ CHashClass *AssignRawTypeToNode(CCmpCtrl *cc,CRPN *r)
-    break;case IC_GET_LABEL:
-    goto unimp;
+    break;case IC_GET_LABEL: return r->ic_class;
--- a/c/aiwn_lexparser.h
+++ b/c/aiwn_lexparser.h
@@ -719,5 +719,6 @@ CCodeMisc *__HC_CodeMiscStrNew(CCmpCtrl *ccmp, char *str, int64_t sz);
 CCodeMisc *__HC_CodeMiscLabelNew(CCmpCtrl *ccmp, void **);
 CRPN *__HC_ICAdd_Goto(CCodeCtrl *cc, CCodeMisc *cm);
 CRPN *__HC_ICAdd_GotoIf(CCodeCtrl *cc, CCodeMisc *cm);
+CRPN *__HC_ICAdd_LabelAddr(CCodeCtrl *cc, CCodeMisc *misc);
 CRPN *__HC_ICAdd_Str(CCodeCtrl *cc, CCodeMisc *cm);
 CRPN *__HC_ICAdd_Label(CCodeCtrl *cc, CCodeMisc *misc);
--- a/c/main.c
+++ b/c/main.c
@@ -1157,5 +1157,8 @@ static int64_t STK___HC_CodeMiscLabelNew(int64_t *stk) {
   return (int64_t)__HC_CodeMiscLabelNew((CCmpCtrl *)stk[0], (void **)stk[1]);
 }
+static int64_t STK___HC_ICAdd_LabelAddr(int64_t *stk) {
+  return (int64_t)__HC_ICAdd_LabelAddr((CCodeCtrl *)stk[0], (CCodeMisc *)stk[1]);
+}
 static int64_t STK___HC_CmpCtrlNew(int64_t *stk) {
   return (int64_t)__HC_CmpCtrlNew();
 }
@@ -1782,4 +1785,5 @@ int main(int argc, char **argv) {
     PrsAddSymbol("__HC_CodeMiscJmpTableNew", STK___HC_CodeMiscJmpTableNew, 4);
     PrsAddSymbol("__HC_ICAdd_Label", STK___HC_ICAdd_Label, 2);
+    PrsAddSymbol("__HC_ICAdd_LabelAddr", STK___HC_ICAdd_LabelAddr, 2);
     PrsAddSymbol("__HC_ICAdd_Goto", STK___HC_ICAdd_Goto, 2);
     PrsAddSymbol("__HC_ICAdd_GotoIf", STK___HC_ICAdd_GotoIf, 2);
--- a/c/parser.c
+++ b/c/parser.c
@@ -4921,5 +4921,17 @@ CRPN *__HC_ICAdd_Label(CCodeCtrl *cc, CCodeMisc *misc) {
   return rpn;
 }
 
+CRPN *__HC_ICAdd_LabelAddr(CCodeCtrl *cc, CCodeMisc *misc) {
+  CRPN *rpn;
+  *(rpn = A_CALLOC(sizeof(CRPN), cc->hc)) = (CRPN){
+      .type = IC_SHORT_ADDR,
+      .code_misc = misc,
+  };
+  rpn->ic_class = HashFind("U8i", Fs->hash_table, HTT_CLASS, 1) + 1;
+  rpn->raw_type = RT_PTR;
+  QueIns(rpn, cc->ir_code);
+  return rpn;
+}
+
 CRPN *__HC_ICAdd_Goto(CCodeCtrl *cc, CCodeMisc *cm) {
   CRPN *rpn;
--- a/c/optpass.c
+++ b/c/optpass.c
@@ -36,7 +36,10 @@ static void FixFunArgs(CCmpCtrl *cctrl, CRPN *rpn) {
   CMemberLst *arg, *args_lst_rev;
   for (cnt = 0; cnt != rpn->length; cnt++)
-    rpn2 = ICFwd(rpn2);
+    if (!(rpn2 = ICFwd(rpn2)))
+      return;
   AssignRawTypeToNode(cctrl, rpn2);
   ic_fun = rpn2->ic_fun;
+  if (!ic_fun)
+    return;
   rpn2 = rpn->base.next;
   arg = ic_fun->base.members_lst;
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -2043,7 +2050,10 @@ CRPN *CopyCode(CCmpCtrl *cc,CCodeCtrl *cc2,CAiwniosAOT *acc)
     new=__HC_ICAdd_AddEq(cc2);
     break;case IC_SUB_EQU:
     new=__HC_ICAdd_SubEq(cc2);
     break;case IC_JMP:
-    new=__HC_ICAdd_Goto(cc2,cur->ic_data(CCodeMisc*)->addr);
+    misc=cur->ic_data;
+    if (!misc->addr)
+      misc->addr=__HC_CodeMiscLabelNew(acc,&misc->addr2);
+    new=__HC_ICAdd_Goto(cc2,misc->addr);
     break;case IC_SUB_RET:
     new=__HC_ICAdd_SubRet(cc2);
