diff --git a/recipe/files/gcc/gcc/config/arm/riscos.opt b/recipe/files/gcc/gcc/config/arm/riscos.opt
--- a/recipe/files/gcc/gcc/config/arm/riscos.opt
+++ b/recipe/files/gcc/gcc/config/arm/riscos.opt
@@ -35,3 +35,7 @@ munixlib
 mmodule
 Target Report RejectNegative Negative(munixlib) Mask(MODULE)
 Target for RISC OS modules
+
+msharedclibrary
+Target Report RejectNegative Mask(SHAREDCLIBRARY)
+Generate code for the RISC OS SharedCLibrary module
diff --git a/recipe/patches/gcc/gcc.config.arm.arm.c.p b/recipe/patches/gcc/gcc.config.arm.arm.c.p
--- a/recipe/patches/gcc/gcc.config.arm.arm.c.p
+++ b/recipe/patches/gcc/gcc.config.arm.arm.c.p
@@ -361,7 +361,7 @@
 +
 +      sl_reg = gen_rtx_REG (Pmode, SL_REGNUM);
 +      sb_ref = gen_rtx_MEM (Pmode,
-+			    gen_rtx_PLUS (Pmode, sl_reg, GEN_INT (-536)));
++                            gen_rtx_PLUS (Pmode, sl_reg, GEN_INT (TARGET_SHAREDCLIBRARY ? -540 : -536)));
 +      MEM_READONLY_P (sb_ref) = 1;
 +      emit_insn (gen_rtx_SET (SImode, pic_reg, sb_ref));
 +    }
