diff --git a/riscos/build-it b/riscos/build-it
--- a/riscos/build-it
+++ b/riscos/build-it
@@ -15,3 +15,4 @@
 #   -f       Force build everything (otherwise skip already built targets)
 #   -static  Skip building shared library related tools
+#   -host-tools-only  Build only compiler host tools
 
@@ -25,7 +26,8 @@
 give_help() {
-  echo -e "$0 [-h|-f|-static] [ \"\" | \"cross\" | \"riscos\" ]\n"
+  echo -e "$0 [-h|-f|-static|-host-tools-only] [ \"\" | \"cross\" | \"riscos\" ]\n"
   echo -e "\t-h\t\tThis help output"
   echo -e "\t-f\t\tForce build everything (otherwise skip already built targets)"
   echo -e "\t-static\t\tDon't build any shared library related tools"
+  echo -e "\t-host-tools-only\tBuild only compiler host tools"
   echo
   echo -e "\tcross\t\tBuild the RISC OS specific binaries suited for cross-compilation usage (default)"
@@ -39,4 +39,5 @@ create_empty_dir() {
 # Setup the default values which can be overruled by user options given:
 GCCSDK_BUILD_FORCED=false
 GCCSDK_BUILD_ISSTATIC=false
+GCCSDK_BUILD_HOST_TOOLS_ONLY=false
 for param in $@ ; do
@@ -48,6 +49,9 @@ for param in $@ ; do
     -f)
        GCCSDK_BUILD_FORCED=true
        ;;
+    -host-tools-only)
+       GCCSDK_BUILD_HOST_TOOLS_ONLY=true
+       ;;
     -static)
        GCCSDK_BUILD_ISSTATIC=true
        ;;
@@ -144,6 +148,10 @@ for roapp in asasm cmunge elf2aif ln mkresfs ; do
   else
     echo +++ NOT building $roapp as it already exists
   fi
 done
+
+if $GCCSDK_BUILD_HOST_TOOLS_ONLY ; then
+  exit 0
+fi
 
 # DRenderer module:
