--- a/Src/AsmResolve.HC
+++ b/Src/AsmResolve.HC
@@ -169,6 +169,8 @@ U0 AOTGlblsResolve(CCmpCtrl *cc,CAOT *tmpaot)
 	      !Bt(&tmpf->flags,Cf_EXTERN)) {
 	  tmpex->val=tmpf->exe_addr;
 	  tmpex->type&=~HTF_UNRESOLVED;
+          AsmExpressionImportsResolve(tmpex->str,tmpu->machine_code,
+                tmpf->exe_addr);
 	  label=tmpauh->hash->str;
 	} else if (tmpex->type&HTT_EXPORT_SYS_SYM &&
 	      (resolved_export=AOTResolvedAsmExportFind(cc,tmpex->str))) {
