diff --git a/Src/PrsVar.HC b/Src/PrsVar.HC
--- a/Src/PrsVar.HC
+++ b/Src/PrsVar.HC
@@ -279,14 +279,8 @@ U0 PrsArrayDims(CCmpCtrl *cc,I64 mode,CArrayDim *dim)
       }
       tmpad=MAlloc(sizeof(CArrayDim));
       tmpad->next=NULL;
-      tmpad1=&dim;
+      tmpad1=dim;
       do {
-      //
-      //Nroot Fix,getting addres of &dim is an 8byte ptr,SO DONT WRITE INTO IT IT
-      // [RBP]
-      // [RBP-8] &dim(writing a member may corrupt RBP)
-      //
-      if(&dim!=tmpad1)
         tmpad1->total_cnt*=j;
 	if (!tmpad1->next) {
 	  tmpad1->next=tmpad;
